// @runware/sdk/dist/index.js
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/types/task-map.ts
var architectureTaskTypes = {
  "exactly-illustrative": "imageInference",
  "flux-1-dev": "imageInference",
  "flux-1-kontext-dev": "imageInference",
  "flux-1-schnell": "imageInference",
  illustrious: "imageInference",
  noobai: "imageInference",
  pony: "imageInference",
  "sd-1-5": "imageInference",
  sdxl: "imageInference",
  "sdxl-lightning": "imageInference",
  "sdxl-turbo": "imageInference",
  "z-image": "imageInference",
  "z-image-turbo": "imageInference"
};
var modalityTaskTypes = {
  "3d": "3dInference",
  audio: "audioInference",
  image: "imageInference",
  text: "textInference",
  video: "videoInference"
};
var operationTaskTypes = {
  caption: "caption",
  "caption-image": "caption",
  "caption-video": "caption",
  "controlnet-preprocess": "controlNetPreprocess",
  masking: "imageMasking",
  "prompt-enhance": "promptEnhance",
  "remove-background": "removeBackground",
  "remove-background-image": "removeBackground",
  "remove-background-video": "removeBackground",
  training: "training",
  upscale: "upscale",
  "upscale-image": "upscale",
  "upscale-video": "upscale",
  vectorize: "vectorize"
};
var models = {
  "alibaba:happyhorse@1.0": { taskType: "videoInference", id: "alibaba-happyhorse-1-0" },
  "alibaba:happyhorse@1.1": { taskType: "videoInference", id: "alibaba-happyhorse-1-1" },
  "alibaba:qwen-image@2.0": { taskType: "imageInference", id: "alibaba-qwen-image-2-0" },
  "alibaba:qwen-image@2.0-pro": { taskType: "imageInference", id: "alibaba-qwen-image-2-0-pro" },
  "alibaba:qwen-image@2512": { taskType: "imageInference", id: "alibaba-qwen-image-2512" },
  "alibaba:qwen-image@layered": { taskType: "imageInference", id: "alibaba-qwen-image-layered" },
  "alibaba:qwen@3-tts-1.7b-base": { taskType: "audioInference", id: "alibaba-qwen3-tts-1-7b-base" },
  "alibaba:qwen@3-tts-1.7b-customvoice": { taskType: "audioInference", id: "alibaba-qwen3-tts-1-7b-customvoice" },
  "alibaba:qwen@3-tts-1.7b-voicedesign": { taskType: "audioInference", id: "alibaba-qwen3-tts-1-7b-voicedesign" },
  "alibaba:qwen@3.5-27b": { taskType: "textInference", id: "alibaba-qwen3-5-27b" },
  "alibaba:qwen@3.5-397b": { taskType: "textInference", id: "alibaba-qwen3-5-397b" },
  "alibaba:wan@2.6": { taskType: "videoInference", id: "alibaba-wan2-6" },
  "alibaba:wan@2.6-flash": { taskType: "videoInference", id: "alibaba-wan2-6-flash" },
  "alibaba:wan@2.6-image": { taskType: "imageInference", id: "alibaba-wan2-6-image" },
  "alibaba:wan@2.7": { taskType: "videoInference", id: "alibaba-wan2-7" },
  "alibaba:wan@2.7-image": { taskType: "imageInference", id: "alibaba-wan2-7-image" },
  "alibaba:wan@2.7-image-pro": { taskType: "imageInference", id: "alibaba-wan2-7-image-pro" },
  "anthropic:claude@fable-5": { taskType: "textInference", id: "anthropic-claude-fable-5" },
  "anthropic:claude@haiku-4.5": { taskType: "textInference", id: "anthropic-claude-haiku-4-5" },
  "anthropic:claude@opus-4.7": { taskType: "textInference", id: "anthropic-claude-opus-4-7" },
  "anthropic:claude@opus-4.8": { taskType: "textInference", id: "anthropic-claude-opus-4-8" },
  "anthropic:claude@sonnet-4.6": { taskType: "textInference", id: "anthropic-claude-sonnet-4-6" },
  "baidu:ernie-image@0": { taskType: "imageInference", id: "baidu-ernie-image" },
  "baidu:ernie-image@turbo": { taskType: "imageInference", id: "baidu-ernie-image-turbo" },
  "bfl:1@2": { taskType: "imageInference", id: "bfl-flux-1-fill-pro" },
  "bfl:1@3": { taskType: "imageInference", id: "bfl-flux-1-expand-pro" },
  "bfl:2@1": { taskType: "imageInference", id: "bfl-flux-1-1-pro" },
  "bfl:2@2": { taskType: "imageInference", id: "bfl-flux-1-1-pro-ultra" },
  "bfl:3@1": { taskType: "imageInference", id: "bfl-flux-1-kontext-pro" },
  "bfl:4@1": { taskType: "imageInference", id: "bfl-flux-1-kontext-max" },
  "bfl:5@1": { taskType: "imageInference", id: "bfl-flux-2-pro" },
  "bfl:6@1": { taskType: "imageInference", id: "bfl-flux-2-flex" },
  "bfl:7@1": { taskType: "imageInference", id: "bfl-flux-2-max" },
  "bfl:flux@erase": { taskType: "imageInference", id: "bfl-flux-erase" },
  "bfl:flux@outpainting": { taskType: "imageInference", id: "bfl-flux-outpainting" },
  "bfl:flux@vto": { taskType: "imageInference", id: "bfl-flux-virtual-try-on" },
  "bria:10@1": { taskType: "imageInference", id: "bria-3-2" },
  "bria:11@1": { taskType: "imageInference", id: "bria-image-replace-background" },
  "bria:2@1": { taskType: "removeBackground", id: "bria-rmbg-v2-0" },
  "bria:20@1": { taskType: "imageInference", id: "bria-fibo" },
  "bria:20@3": { taskType: "imageInference", id: "bria-fibo-lite" },
  "bria:21@1": { taskType: "imageInference", id: "bria-fibo-edit" },
  "bria:21@2": { taskType: "imageInference", id: "bria-fibo-edit-tools" },
  "bria:50@1": { taskType: "upscale", id: "bria-video-increase-resolution" },
  "bria:51@1": { taskType: "removeBackground", id: "bria-video-background-removal" },
  "bria:52@1": { taskType: "upscale", id: "bria-image-increase-resolution" },
  "bria:60@1": { taskType: "videoInference", id: "bria-video-eraser" },
  "bytedance:2@1": { taskType: "videoInference", id: "bytedance-seedance-1-0-pro" },
  "bytedance:2@2": { taskType: "videoInference", id: "bytedance-seedance-1-0-pro-fast" },
  "bytedance:5@0": { taskType: "imageInference", id: "bytedance-seedream-4-0" },
  "bytedance:5@1": { taskType: "videoInference", id: "bytedance-omnihuman-1" },
  "bytedance:5@2": { taskType: "videoInference", id: "bytedance-omnihuman-1-5" },
  "bytedance:50@1": { taskType: "upscale", id: "bytedance-video-upscaler" },
  "bytedance:seed-audio@1.0": { taskType: "audioInference", id: "bytedance-seed-audio-1-0" },
  "bytedance:seedance@1.5-pro": { taskType: "videoInference", id: "bytedance-seedance-1-5-pro" },
  "bytedance:seedance@2.0": { taskType: "videoInference", id: "bytedance-seedance-2-0" },
  "bytedance:seedance@2.0-fast": { taskType: "videoInference", id: "bytedance-seedance-2-0-fast" },
  "bytedance:seedance@2.0-mini": { taskType: "videoInference", id: "bytedance-seedance-2-0-mini" },
  "bytedance:seedream@4.5": { taskType: "imageInference", id: "bytedance-seedream-4-5" },
  "bytedance:seedream@5.0-lite": { taskType: "imageInference", id: "bytedance-seedream-5-0-lite" },
  "bytedance:seedream@5.0-pro": { taskType: "imageInference", id: "bytedance-seedream-5-0-pro" },
  "bytedance:video-enhancement@pro": { taskType: "upscale", id: "bytedance-video-enhancement-pro" },
  "bytedance:video-enhancement@standard": { taskType: "upscale", id: "bytedance-video-enhancement-standard" },
  "civitai:101055@128078": { taskType: "imageInference", id: "stabilityai-stable-diffusion-xl-v1-0-vae-fix" },
  "creatify:aurora@0": { taskType: "videoInference", id: "creatify-aurora-v1" },
  "creatify:aurora@fast": { taskType: "videoInference", id: "creatify-aurora-v1-fast" },
  "deepseek:v4@flash": { taskType: "textInference", id: "deepseek-v4-flash" },
  "deepseek:v4@pro": { taskType: "textInference", id: "deepseek-v4-pro" },
  "exactly:illustrative@training": { taskType: "training", id: "exactly-illustrative-training" },
  "exactly:photo@bright-pulse": { taskType: "imageInference", id: "exactly-photo-bright-pulse" },
  "exactly:photo@distant-reality": { taskType: "imageInference", id: "exactly-photo-distant-reality" },
  "exactly:photo@extreme-contrast": { taskType: "imageInference", id: "exactly-photo-extreme-contrast" },
  "exactly:photo@grain-film-look": { taskType: "imageInference", id: "exactly-photo-grain-film-look" },
  "exactly:photo@journey": { taskType: "imageInference", id: "exactly-photo-journey" },
  "exactly:photo@warm-light": { taskType: "imageInference", id: "exactly-photo-warm-light" },
  "fishaudio:s2.1@pro": { taskType: "audioInference", id: "fish-audio-s2-1-pro" },
  "google:3@2": { taskType: "videoInference", id: "google-veo-3-1" },
  "google:3@3": { taskType: "videoInference", id: "google-veo-3-1-fast" },
  "google:4@1": { taskType: "imageInference", id: "google-nano-banana" },
  "google:4@2": { taskType: "imageInference", id: "google-nano-banana-pro" },
  "google:4@3": { taskType: "imageInference", id: "google-nano-banana-2" },
  "google:gemini@3-flash": { taskType: "textInference", id: "google-gemini-3-flash" },
  "google:gemini@3.1-flash-lite": { taskType: "textInference", id: "google-gemini-3-1-flash-lite" },
  "google:gemini@3.1-flash-tts": { taskType: "audioInference", id: "google-gemini-3-1-flash-tts" },
  "google:gemini@3.1-pro": { taskType: "textInference", id: "google-gemini-3-1-pro" },
  "google:gemini@3.5-flash": { taskType: "textInference", id: "google-gemini-3-5-flash" },
  "google:gemini@omni-flash": { taskType: "videoInference", id: "google-gemini-omni-flash" },
  "google:gemma@4-31b": { taskType: "textInference", id: "google-gemma-4-31b" },
  "google:nano-banana@2-lite": { taskType: "imageInference", id: "google-nano-banana-2-lite" },
  "google:veo@3.1-lite": { taskType: "videoInference", id: "google-veo-3-1-lite" },
  "heygen:avatar@4": { taskType: "videoInference", id: "heygen-avatar-iv" },
  "heygen:avatar@5": { taskType: "videoInference", id: "heygen-avatar-v" },
  "heygen:video-agent@0": { taskType: "videoInference", id: "heygen-video-agent" },
  "hyper3d:rodin@gen-2": { taskType: "3dInference", id: "hyper3d-rodin-gen-2" },
  "ideogram:2@1": { taskType: "imageInference", id: "ideogram-2a" },
  "ideogram:2@2": { taskType: "imageInference", id: "ideogram-2a-remix" },
  "ideogram:3@1": { taskType: "imageInference", id: "ideogram-2-0" },
  "ideogram:3@2": { taskType: "imageInference", id: "ideogram-2-0-remix" },
  "ideogram:4@0": { taskType: "imageInference", id: "ideogram-4-0" },
  "ideogram:4@1": { taskType: "imageInference", id: "ideogram-3-0" },
  "ideogram:4@2": { taskType: "imageInference", id: "ideogram-3-0-remix" },
  "ideogram:4@3": { taskType: "imageInference", id: "ideogram-3-0-edit" },
  "ideogram:4@4": { taskType: "imageInference", id: "ideogram-3-0-reframe" },
  "ideogram:4@5": { taskType: "imageInference", id: "ideogram-3-0-replace-background" },
  "ideogram:4@remix": { taskType: "imageInference", id: "ideogram-4-0-remix" },
  "ideogram:layerize-text@0": { taskType: "imageInference", id: "ideogram-layerize-text" },
  "imagineart:1.5-pro@0": { taskType: "imageInference", id: "imagineart-1-5-pro" },
  "imagineart:1@5": { taskType: "imageInference", id: "imagineart-1-5" },
  "imagineart:2.0@0": { taskType: "imageInference", id: "imagineart-2-0" },
  "inworld:tts@1.5-max": { taskType: "audioInference", id: "inworld-tts-1-5-max" },
  "inworld:tts@1.5-mini": { taskType: "audioInference", id: "inworld-tts-1-5-mini" },
  "inworld:tts@2": { taskType: "audioInference", id: "inworld-tts-2" },
  "klingai:1@1": { taskType: "videoInference", id: "klingai-1-0-standard" },
  "klingai:1@2": { taskType: "videoInference", id: "klingai-1-0-pro" },
  "klingai:2@1": { taskType: "videoInference", id: "klingai-1-5-standard" },
  "klingai:2@2": { taskType: "videoInference", id: "klingai-1-5-pro" },
  "klingai:3@1": { taskType: "videoInference", id: "klingai-1-6-standard" },
  "klingai:3@2": { taskType: "videoInference", id: "klingai-1-6-pro" },
  "klingai:4@3": { taskType: "videoInference", id: "klingai-2-0-master" },
  "klingai:5@1": { taskType: "videoInference", id: "klingai-2-1-standard" },
  "klingai:5@2": { taskType: "videoInference", id: "klingai-2-1-pro" },
  "klingai:5@3": { taskType: "videoInference", id: "klingai-2-1-master" },
  "klingai:6@0": { taskType: "videoInference", id: "klingai-2-5-turbo-standard" },
  "klingai:6@1": { taskType: "videoInference", id: "klingai-2-5-turbo-pro" },
  "klingai:7@1": { taskType: "videoInference", id: "klingai-lip-sync" },
  "klingai:avatar@2.0-pro": { taskType: "videoInference", id: "klingai-avatar-2-0-pro" },
  "klingai:avatar@2.0-standard": { taskType: "videoInference", id: "klingai-avatar-2-0-standard" },
  "klingai:kling-image@3": { taskType: "imageInference", id: "klingai-image-3-0" },
  "klingai:kling-image@o1": { taskType: "imageInference", id: "klingai-image-o1" },
  "klingai:kling-image@o3": { taskType: "imageInference", id: "klingai-image-o3" },
  "klingai:kling-video@2.6-pro": { taskType: "videoInference", id: "klingai-video-2-6-pro" },
  "klingai:kling-video@2.6-standard": { taskType: "videoInference", id: "klingai-video-2-6-standard" },
  "klingai:kling-video@3-4k": { taskType: "videoInference", id: "klingai-video-3-0-4k" },
  "klingai:kling-video@3-pro": { taskType: "videoInference", id: "klingai-video-3-0-pro" },
  "klingai:kling-video@3-standard": { taskType: "videoInference", id: "klingai-video-3-0-standard" },
  "klingai:kling-video@3.0-turbo": { taskType: "videoInference", id: "klingai-video-3-0-turbo" },
  "klingai:kling-video@o3-4k": { taskType: "videoInference", id: "klingai-video-3-0-omni-4k" },
  "klingai:kling-video@o3-pro": { taskType: "videoInference", id: "klingai-video-3-0-omni-pro" },
  "klingai:kling-video@o3-standard": { taskType: "videoInference", id: "klingai-video-3-0-omni-standard" },
  "klingai:kling@o1": { taskType: "videoInference", id: "klingai-video-o1-pro" },
  "klingai:kling@o1-standard": { taskType: "videoInference", id: "klingai-video-o1-standard" },
  "krea:krea@2-large": { taskType: "imageInference", id: "krea-2-large" },
  "krea:krea@2-medium": { taskType: "imageInference", id: "krea-2-medium" },
  "krea:krea@2-turbo": { taskType: "imageInference", id: "krea-2-turbo" },
  "lightricks:2@0": { taskType: "videoInference", id: "lightricks-ltx-2-pro" },
  "lightricks:2@1": { taskType: "videoInference", id: "lightricks-ltx-2-fast" },
  "lightricks:3@1": { taskType: "videoInference", id: "lightricks-ltx-2-retake" },
  "lightricks:ltx@2": { taskType: "videoInference", id: "lightricks-ltx-2" },
  "lightricks:ltx@2.3": { taskType: "videoInference", id: "lightricks-ltx-2-3" },
  "lightricks:ltx@2.3-fast": { taskType: "videoInference", id: "lightricks-ltx-2-3-fast" },
  "luma:ray@3.2": { taskType: "videoInference", id: "luma-ray3-2" },
  "luma:uni@1": { taskType: "imageInference", id: "luma-uni-1" },
  "luma:uni@1-max": { taskType: "imageInference", id: "luma-uni-1-max" },
  "memories:1@1": { taskType: "caption", id: "memories-video-captioning" },
  "memories:2@1": { taskType: "caption", id: "memories-video-age-detection" },
  "meshy:meshy@6": { taskType: "3dInference", id: "meshy-6" },
  "meta:sam@3d": { taskType: "3dInference", id: "meta-sam-3d-objects" },
  "microsoft:trellis-2@4b": { taskType: "3dInference", id: "microsoft-trellis-2" },
  "minimax:1@1": { taskType: "videoInference", id: "minimax-01" },
  "minimax:2@1": { taskType: "videoInference", id: "minimax-01-director" },
  "minimax:2@3": { taskType: "videoInference", id: "minimax-01-live" },
  "minimax:3@1": { taskType: "videoInference", id: "minimax-hailuo-02" },
  "minimax:4@1": { taskType: "videoInference", id: "minimax-hailuo-2-3" },
  "minimax:4@2": { taskType: "videoInference", id: "minimax-hailuo-2-3-fast" },
  "minimax:m2.5@0": { taskType: "textInference", id: "minimax-m2-5" },
  "minimax:m2.7@0": { taskType: "textInference", id: "minimax-m2-7" },
  "minimax:m2.7@highspeed": { taskType: "textInference", id: "minimax-m2-7-highspeed" },
  "minimax:m3@0": { taskType: "textInference", id: "minimax-m3" },
  "minimax:music@2.6": { taskType: "audioInference", id: "minimax-music-2-6" },
  "minimax:music@cover": { taskType: "audioInference", id: "minimax-music-cover" },
  "minimax:speech@2.8": { taskType: "audioInference", id: "minimax-speech-2-8" },
  "mirelo:1@1": { taskType: "audioInference", id: "mirelo-sfx-1-5" },
  "mirelo:sfx@1.6": { taskType: "audioInference", id: "mirelo-sfx-1-6" },
  "moonshotai:kimi@k2.6": { taskType: "textInference", id: "moonshotai-kimi-k2-6" },
  "openai:1@1": { taskType: "imageInference", id: "openai-gpt-image-1" },
  "openai:1@2": { taskType: "imageInference", id: "openai-gpt-image-1-mini" },
  "openai:3@1": { taskType: "videoInference", id: "openai-sora-2" },
  "openai:3@2": { taskType: "videoInference", id: "openai-sora-2-pro" },
  "openai:4@1": { taskType: "imageInference", id: "openai-gpt-image-1-5" },
  "openai:gpt-image@2": { taskType: "imageInference", id: "openai-gpt-image-2" },
  "openai:gpt@5-mini": { taskType: "textInference", id: "openai-gpt-5-mini" },
  "openai:gpt@5-nano": { taskType: "textInference", id: "openai-gpt-5-nano" },
  "openai:gpt@5.4": { taskType: "textInference", id: "openai-gpt-5-4" },
  "openai:gpt@5.4-mini": { taskType: "textInference", id: "openai-gpt-5-4-mini" },
  "openai:gpt@5.4-nano": { taskType: "textInference", id: "openai-gpt-5-4-nano" },
  "openai:gpt@5.4-pro": { taskType: "textInference", id: "openai-gpt-5-4-pro" },
  "openai:gpt@5.5": { taskType: "textInference", id: "openai-gpt-5-5" },
  "picsart:1@1": { taskType: "vectorize", id: "picsart-image-vectorizer" },
  "pixverse:1@1": { taskType: "videoInference", id: "pixverse-v3-5" },
  "pixverse:1@2": { taskType: "videoInference", id: "pixverse-v4" },
  "pixverse:1@3": { taskType: "videoInference", id: "pixverse-v4-5" },
  "pixverse:1@5": { taskType: "videoInference", id: "pixverse-v5" },
  "pixverse:1@5-fast": { taskType: "videoInference", id: "pixverse-v5-fast" },
  "pixverse:1@6": { taskType: "videoInference", id: "pixverse-v5-5" },
  "pixverse:1@7": { taskType: "videoInference", id: "pixverse-v5-6" },
  "pixverse:1@8": { taskType: "videoInference", id: "pixverse-v6" },
  "pixverse:lipsync@1": { taskType: "videoInference", id: "pixverse-lipsync" },
  "pixverse:modify@0": { taskType: "videoInference", id: "pixverse-modify" },
  "prunaai:1@1": { taskType: "imageInference", id: "prunaai-p-image" },
  "prunaai:2@1": { taskType: "imageInference", id: "prunaai-p-image-edit" },
  "prunaai:p-image@try-on": { taskType: "imageInference", id: "prunaai-p-image-try-on" },
  "prunaai:p-image@upscale": { taskType: "upscale", id: "prunaai-p-image-upscale" },
  "prunaai:p-video@0": { taskType: "videoInference", id: "prunaai-p-video" },
  "prunaai:p-video@animate": { taskType: "videoInference", id: "prunaai-p-video-animate" },
  "prunaai:p-video@avatar": { taskType: "videoInference", id: "prunaai-p-video-avatar" },
  "prunaai:p-video@replace": { taskType: "videoInference", id: "prunaai-p-video-replace" },
  "recraft:1@1": { taskType: "vectorize", id: "recraft-vectorize" },
  "recraft:v4-pro@0": { taskType: "imageInference", id: "recraft-v4-pro" },
  "recraft:v4-pro@vector": { taskType: "vectorize", id: "recraft-v4-pro-vector" },
  "recraft:v4.1-pro@0": { taskType: "imageInference", id: "recraft-v4-1-pro" },
  "recraft:v4.1-utility-pro@0": { taskType: "imageInference", id: "recraft-v4-1-utility-pro" },
  "recraft:v4.1-utility@0": { taskType: "imageInference", id: "recraft-v4-1-utility" },
  "recraft:v4.1@0": { taskType: "imageInference", id: "recraft-v4-1" },
  "recraft:v4@0": { taskType: "imageInference", id: "recraft-v4" },
  "recraft:v4@vector": { taskType: "vectorize", id: "recraft-v4-vector" },
  "rundiffusion:110@101": { taskType: "imageInference", id: "rundiffusion-juggernaut-lightning-flux" },
  "rundiffusion:120@100": { taskType: "imageInference", id: "rundiffusion-juggernaut-base-flux" },
  "rundiffusion:130@100": { taskType: "imageInference", id: "rundiffusion-juggernaut-pro-flux" },
  "rundiffusion:200@100": { taskType: "imageInference", id: "rundiffusion-juggernaut-z" },
  "runware:100@1": { taskType: "imageInference", id: "bfl-flux-1-schnell" },
  "runware:101@1": { taskType: "imageInference", id: "bfl-flux-1-dev" },
  "runware:102@1": { taskType: "imageInference", id: "bfl-flux-1-fill-dev" },
  "runware:106@1": { taskType: "imageInference", id: "bfl-flux-1-kontext-dev" },
  "runware:107@1": { taskType: "imageInference", id: "krea-flux-1-krea-dev" },
  "runware:108@1": { taskType: "imageInference", id: "alibaba-qwen-image" },
  "runware:108@20": { taskType: "imageInference", id: "alibaba-qwen-image-edit" },
  "runware:108@22": { taskType: "imageInference", id: "alibaba-qwen-image-edit-plus" },
  "runware:109@1": { taskType: "removeBackground", id: "rembg-v1-4" },
  "runware:110@1": { taskType: "removeBackground", id: "bria-rmbg-v2-0-open" },
  "runware:111@1": { taskType: "imageInference", id: "flux-1-dev-srpo" },
  "runware:112@1": { taskType: "removeBackground", id: "birefnet-v1-base" },
  "runware:112@10": { taskType: "removeBackground", id: "birefnet-portrait" },
  "runware:112@2": { taskType: "removeBackground", id: "birefnet-v1-base-cod" },
  "runware:112@3": { taskType: "removeBackground", id: "birefnet-dis" },
  "runware:112@5": { taskType: "removeBackground", id: "birefnet-general" },
  "runware:112@6": { taskType: "removeBackground", id: "birefnet-general-resolution-512x512-fp16" },
  "runware:112@7": { taskType: "removeBackground", id: "birefnet-hrsod-dhu" },
  "runware:112@8": { taskType: "removeBackground", id: "birefnet-massive-tr-dis5k-tr-tes" },
  "runware:112@9": { taskType: "removeBackground", id: "birefnet-matting" },
  "runware:150@2": { taskType: "caption", id: "meta-llava-1-6-mistral-7b" },
  "runware:151@1": { taskType: "caption", id: "openai-clip-vit-l-14" },
  "runware:152@1": { taskType: "caption", id: "alibaba-qwen2-5-vl-3b-instruct" },
  "runware:152@2": { taskType: "caption", id: "alibaba-qwen2-5-vl-7b-instruct" },
  "runware:152@50": { taskType: "caption", id: "qwen2-5-vl-7b-age-detector" },
  "runware:153@1": { taskType: "caption", id: "vit-age-classifier" },
  "runware:154@1": { taskType: "caption", id: "open-age-detection" },
  "runware:180@1": { taskType: "imageInference", id: "tencent-hunyuanimage-3-0" },
  "runware:190@1": { taskType: "videoInference", id: "ovi" },
  "runware:201@1": { taskType: "videoInference", id: "alibaba-wan2-5-preview" },
  "runware:201@10": { taskType: "imageInference", id: "alibaba-wan2-5-preview-image" },
  "runware:300@1": { taskType: "imageInference", id: "object-eraser" },
  "runware:35@1": { taskType: "imageMasking", id: "yolov8n-face" },
  "runware:35@10": { taskType: "imageMasking", id: "mediapipe-eyes-lips-mesh" },
  "runware:35@11": { taskType: "imageMasking", id: "mediapipe-nose-eyes-mesh" },
  "runware:35@12": { taskType: "imageMasking", id: "mediapipe-nose-lips-mesh" },
  "runware:35@13": { taskType: "imageMasking", id: "mediapipe-nose-mesh" },
  "runware:35@14": { taskType: "imageMasking", id: "mediapipe-lips-mesh" },
  "runware:35@15": { taskType: "imageMasking", id: "mediapipe-eyes-mesh" },
  "runware:35@2": { taskType: "imageMasking", id: "yolov8s-face" },
  "runware:35@3": { taskType: "imageMasking", id: "yolov8n-hand" },
  "runware:35@4": { taskType: "imageMasking", id: "yolov8n-person-seg" },
  "runware:35@5": { taskType: "imageMasking", id: "yolov8s-person-seg" },
  "runware:35@6": { taskType: "imageMasking", id: "mediapipe-face-full" },
  "runware:35@7": { taskType: "imageMasking", id: "mediapipe-face-short" },
  "runware:35@8": { taskType: "imageMasking", id: "mediapipe-face-mesh" },
  "runware:35@9": { taskType: "imageMasking", id: "mediapipe-face-mesh-eyes-only" },
  "runware:400@1": { taskType: "imageInference", id: "bfl-flux-2-dev" },
  "runware:400@2": { taskType: "imageInference", id: "bfl-flux-2-klein-9b" },
  "runware:400@3": { taskType: "imageInference", id: "bfl-flux-2-klein-9b-base" },
  "runware:400@4": { taskType: "imageInference", id: "bfl-flux-2-klein-4b" },
  "runware:400@5": { taskType: "imageInference", id: "bfl-flux-2-klein-4b-base" },
  "runware:400@6": { taskType: "imageInference", id: "bfl-flux-2-klein-9b-kv" },
  "runware:5@1": { taskType: "imageInference", id: "stabilityai-stable-diffusion-3" },
  "runware:500@1": { taskType: "upscale", id: "clarity" },
  "runware:501@1": { taskType: "upscale", id: "ccsr" },
  "runware:502@1": { taskType: "upscale", id: "stable-diffusion-latent-upscaler" },
  "runware:503@1": { taskType: "upscale", id: "swinir" },
  "runware:504@1": { taskType: "upscale", id: "real-esrgan" },
  "runware:97@1": { taskType: "imageInference", id: "hidream-i1-full" },
  "runware:97@2": { taskType: "imageInference", id: "hidream-i1-dev" },
  "runware:97@3": { taskType: "imageInference", id: "hidream-i1-fast" },
  "runware:ace-step@v1.5-base": { taskType: "audioInference", id: "ace-step-v1-5-base" },
  "runware:ace-step@v1.5-turbo": { taskType: "audioInference", id: "ace-step-v1-5-turbo" },
  "runware:ace-step@v1.5-xl-base": { taskType: "audioInference", id: "ace-step-v1-5-xl-base" },
  "runware:ace-step@v1.5-xl-sft": { taskType: "audioInference", id: "ace-step-v1-5-xl-sft" },
  "runware:ace-step@v1.5-xl-turbo": { taskType: "audioInference", id: "ace-step-v1-5-xl-turbo" },
  "runware:controlnet-preprocess@canny": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-canny" },
  "runware:controlnet-preprocess@depth": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-depth" },
  "runware:controlnet-preprocess@lineart": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-lineart" },
  "runware:controlnet-preprocess@lineart-anime": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-lineart-anime" },
  "runware:controlnet-preprocess@mlsd": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-mlsd" },
  "runware:controlnet-preprocess@normalbae": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-normalbae" },
  "runware:controlnet-preprocess@openpose": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-openpose" },
  "runware:controlnet-preprocess@scribble": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-scribble" },
  "runware:controlnet-preprocess@seg": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-seg" },
  "runware:controlnet-preprocess@shuffle": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-shuffle" },
  "runware:controlnet-preprocess@softedge": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-softedge" },
  "runware:controlnet-preprocess@tile": { taskType: "controlNetPreprocess", id: "controlnet-preprocess-tile" },
  "runware:dia@1.6b": { taskType: "audioInference", id: "dia-1-6b" },
  "runware:dia2@2b": { taskType: "audioInference", id: "dia2-2b" },
  "runware:flux-1-dev@style-lora-training": { taskType: "training", id: "runware-flux-1-dev-style-lora-training" },
  "runware:kandinsky-5.0-image-lite@1": { taskType: "imageInference", id: "kandinsky-5-0-image-lite" },
  "runware:llama-3-1-8b@prompt-enhancer": { taskType: "promptEnhance", id: "llama-3-1-8b-prompt-enhancer" },
  "runware:qwen-image@style-lora-training": { taskType: "training", id: "runware-qwen-image-style-lora-training" },
  "runware:twinflow-z-image-turbo@0": { taskType: "imageInference", id: "twinflow-z-image-turbo" },
  "runware:z-image@0": { taskType: "imageInference", id: "alibaba-z-image" },
  "runware:z-image@style-lora-training": { taskType: "training", id: "runware-z-image-style-lora-training" },
  "runware:z-image@turbo": { taskType: "imageInference", id: "alibaba-z-image-turbo" },
  "runway:1@1": { taskType: "videoInference", id: "runway-gen-4-turbo" },
  "runway:1@2": { taskType: "videoInference", id: "runway-gen-4-5" },
  "runway:2@1": { taskType: "videoInference", id: "runway-aleph" },
  "runway:4@1": { taskType: "imageInference", id: "runway-gen-4-image" },
  "runway:4@2": { taskType: "imageInference", id: "runway-gen-4-image-turbo" },
  "runway:aleph@2.0": { taskType: "videoInference", id: "runway-aleph-2-0" },
  "skywork:skyreels@v4": { taskType: "videoInference", id: "skywork-skyreels-v4" },
  "sourceful:riverflow-2.0@fast": { taskType: "imageInference", id: "sourceful-riverflow-2-0-fast" },
  "sourceful:riverflow-2.0@pro": { taskType: "imageInference", id: "sourceful-riverflow-2-0-pro" },
  "sourceful:riverflow-2.5@fast": { taskType: "imageInference", id: "sourceful-riverflow-2-5-fast" },
  "sourceful:riverflow-2.5@pro": { taskType: "imageInference", id: "sourceful-riverflow-2-5-pro" },
  "sync:lipsync-2-pro@1": { taskType: "videoInference", id: "sync-lipsync-2-pro" },
  "sync:lipsync-2@1": { taskType: "videoInference", id: "sync-lipsync-2" },
  "sync:lipsync@3": { taskType: "videoInference", id: "sync-3" },
  "sync:react-1@1": { taskType: "videoInference", id: "sync-react-1" },
  "tencent:hunyuan-3d@3.1-pro": { taskType: "3dInference", id: "tencent-hunyuan-3d-3-1-pro" },
  "tencent:hunyuan-3d@3.1-rapid": { taskType: "3dInference", id: "tencent-hunyuan-3d-3-1-rapid" },
  "topazlabs:starlight-precise@2.5": { taskType: "upscale", id: "topazlabs-starlight-precise-2-5" },
  "tripo:v3.1@0": { taskType: "3dInference", id: "tripo-v3-1" },
  "veed:fabric@1.0-hosted": { taskType: "videoInference", id: "veed-fabric-1-0-hosted" },
  "vidu:1@0": { taskType: "videoInference", id: "vidu-q1-classic" },
  "vidu:1@1": { taskType: "videoInference", id: "vidu-q1" },
  "vidu:1@5": { taskType: "videoInference", id: "vidu-1-5" },
  "vidu:2@0": { taskType: "videoInference", id: "vidu-2-0" },
  "vidu:3@1": { taskType: "videoInference", id: "vidu-q2-pro" },
  "vidu:3@2": { taskType: "videoInference", id: "vidu-q2-turbo" },
  "vidu:4@1": { taskType: "videoInference", id: "vidu-q3" },
  "vidu:4@2": { taskType: "videoInference", id: "vidu-q3-turbo" },
  "vidu:q1@image": { taskType: "imageInference", id: "vidu-q1-image" },
  "xai:grok-imagine@image": { taskType: "imageInference", id: "xai-grok-imagine-image" },
  "xai:grok-imagine@image-quality": { taskType: "imageInference", id: "xai-grok-imagine-image-quality" },
  "xai:grok-imagine@video": { taskType: "videoInference", id: "xai-grok-imagine-video" },
  "xai:grok-imagine@video-1.5": { taskType: "videoInference", id: "xai-grok-imagine-video-1-5" },
  "xai:grok@4.3": { taskType: "textInference", id: "xai-grok-4-3" },
  "xai:tts@0": { taskType: "audioInference", id: "xai-tts" },
  "zai:glm@4.7": { taskType: "textInference", id: "zai-glm-4-7" },
  "zai:glm@5.1": { taskType: "textInference", id: "zai-glm-5-1" }
};

// src/constants.ts
var SCHEMAS_BASE_URL = "https://schemas.runware.ai";
var CONTENT_BASE_URL = "https://content.runware.ai";

// src/validate.ts
var ajvInstance = null;
var ajvLoadPromise = null;
var validatorCache = new Map;
var docsUrlCache = new Map;
var getDocsUrlForModel = (model) => docsUrlCache.get(model) ?? null;
var loadAjv = async () => {
  if (ajvInstance) {
    return;
  }
  if (!ajvLoadPromise) {
    ajvLoadPromise = (async () => {
      try {
        const ajvModule = "ajv/dist/2019.js";
        const ajvMod = await import(ajvModule);
        const Ajv = ajvMod.default;
        ajvInstance = new Ajv({ allErrors: true, strict: false, validateFormats: false });
      } catch {
        ajvLoadPromise = null;
        throw new Error('Validation requires "ajv". Install it: npm install ajv');
      }
    })();
  }
  await ajvLoadPromise;
};
var fetchModelSchema = async (air, config) => {
  const fetchImpl = config.dependencies?.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    return null;
  }
  const url = `${SCHEMAS_BASE_URL}/resolve/${encodeURIComponent(air)}`;
  try {
    const response = await fetchImpl(url);
    if (response.status === 404) {
      return null;
    }
    if (!response.ok) {
      config.log.warn(`Schema resolve failed for ${air}: HTTP ${response.status}`);
      return null;
    }
    const data = await response.json();
    if (data.documentation) {
      docsUrlCache.set(air, data.documentation);
    }
    return data.requestSchema;
  } catch (error) {
    config.log.warn(`Schema resolve error for ${air}`, error);
    return null;
  }
};
var getValidator = async (model, config) => {
  const cached = validatorCache.get(model);
  if (cached) {
    return cached;
  }
  const promise = (async () => {
    await loadAjv();
    const schema = await fetchModelSchema(model, config);
    if (!schema) {
      return null;
    }
    return ajvInstance.compile(schema);
  })();
  validatorCache.set(model, promise);
  try {
    return await promise;
  } catch (err) {
    validatorCache.delete(model);
    throw err;
  }
};
var validateTasks = async (tasks, config) => {
  for (const task of tasks) {
    const model = task.model;
    if (!model) {
      continue;
    }
    const validator = await getValidator(model, config);
    if (!validator) {
      continue;
    }
    if (!validator(task)) {
      const errors = validator.errors ?? [];
      const first = errors[0];
      const parameter = first?.instancePath ? first.instancePath.slice(1).replace(/\//g, ".") : undefined;
      const summary = errors.map((e) => {
        const path = e.instancePath ? e.instancePath.slice(1).replace(/\//g, ".") : "";
        return path ? `${path} ${e.message ?? "is invalid"}` : e.message ?? "is invalid";
      }).join("; ");
      throw createRunwareError("validationFailed", `Validation failed for ${task.taskType}: ${summary}`, {
        taskType: task.taskType,
        taskUUID: task.taskUUID,
        parameter,
        model,
        validationErrors: errors
      });
    }
  }
};
var clearValidatorCache = () => {
  validatorCache.clear();
  docsUrlCache.clear();
};

// src/errors.ts
var DOCS_BASE = "https://runware.ai/docs";
var UTILITY_DOC_PATHS = {
  modelSearch: "platform/model-search",
  modelUpload: "platform/model-upload",
  imageUpload: "platform/media-storage",
  mediaStorage: "platform/media-storage",
  getResponse: "platform/task-polling",
  accountManagement: "platform/account-management"
};
var SDK_ERROR_DOC_PATH = "platform/errors";
var SDK_ONLY_RAW_CODES = new Set([
  "aborted",
  "unknownModel",
  "noFetchImpl",
  "noWebSocketImpl",
  "streamUnsupported",
  "notConnected",
  "notOpen",
  "reconnectionFailed",
  "sendFailed",
  "connectionFailed",
  "missingApiKey",
  "parseError"
]);
var SAFETY_CODES = new Set([
  "contentPolicyViolation",
  "providerContentPolicyViolation",
  "sensitiveContentDetected",
  "unsafeContentDetected",
  "nsfwContentDetected",
  "promptBlocked",
  "imageBlocked",
  "moderationFailed"
]);
var AUTH_CODES = new Set([
  "unauthorized",
  "forbidden",
  "permissionDenied",
  "insufficientPermissions",
  "authenticationFailed",
  "authFailed",
  "authTimeout",
  "invalidAuthentication",
  "invalidCredentials",
  "missingAuthentication",
  "tokenExpired",
  "tokenInvalid",
  "tokenMissing",
  "tokenRevoked",
  "accountSuspended",
  "accountDisabled",
  "organizationSuspended",
  "organizationDisabled",
  "workspaceSuspended",
  "workspaceDisabled"
]);
var SERVER_ERROR_CODES = new Set([
  "internalServerError",
  "serviceUnavailable",
  "serverUnavailable",
  "standardError",
  "unknownError",
  "undefinedError",
  "defaultError",
  "unrecognizedResponse",
  "errorRetrievingAccountManagement"
]);
var NOT_FOUND_CODES = new Set([
  "taskCancelled",
  "taskFailedOrNotFound",
  "unknownModel"
]);
var PROVIDER_CODES = new Set([
  "inferenceError",
  "processingFailed",
  "taskFailed",
  "downloadFailed",
  "uploadFailed",
  "noAvailableServer",
  "modelUnavailable",
  "modelDisabled",
  "modelNotReady",
  "mediaStorageFileCouldNotBeMoved"
]);
var RETRYABLE_CODES = new Set([
  "provider",
  "timeout",
  "connection",
  "rateLimit",
  "serverError"
]);
var isRetryable = (code) => RETRYABLE_CODES.has(code);
var deriveCode = (raw) => {
  if (raw === "aborted") {
    return "aborted";
  }
  if (raw === "connectionFailed" || raw === "notConnected" || raw === "notOpen" || raw === "reconnectionFailed" || raw === "disconnected") {
    return "connection";
  }
  if (raw.includes("Credits") || raw.includes("Quota") || raw.includes("Balance") || raw === "quotaExceeded" || raw === "paymentRequired") {
    return "quota";
  }
  if (raw.includes("RateLimit") || raw === "rateLimitExceeded") {
    return "rateLimit";
  }
  if (SAFETY_CODES.has(raw)) {
    return "safety";
  }
  if (AUTH_CODES.has(raw) || raw.includes("ApiKey")) {
    return "auth";
  }
  if (raw.includes("Timeout") || raw === "timeout") {
    return "timeout";
  }
  if (NOT_FOUND_CODES.has(raw) || raw.endsWith("NotFound") || raw.endsWith("Expired")) {
    return "notFound";
  }
  if (SERVER_ERROR_CODES.has(raw)) {
    return "serverError";
  }
  if (PROVIDER_CODES.has(raw) || raw.startsWith("provider")) {
    return "provider";
  }
  if (raw.startsWith("invalid") || raw.startsWith("missing") || raw.startsWith("conflict") || raw.startsWith("duplicate") || raw.startsWith("unsupported") || raw.startsWith("value") || raw.startsWith("array") || raw === "unknownParameter" || raw === "incompatibleParameters" || raw === "mismatchProviderSettingsProvider" || raw === "unknownProviderSettingsProvider" || raw === "transparentModelMismatch" || raw === "modelOwnershipValidationError" || raw === "modelAlreadyExists" || raw.startsWith("max") || raw === "validationFailed") {
    return "validation";
  }
  return "unknown";
};

class RunwareError extends Error {
  code;
  retryable;
  constructor(rawCode, message) {
    super(message);
    this.name = "RunwareError";
    this.code = deriveCode(rawCode);
    this.retryable = isRetryable(this.code);
  }
}
var paramAnchor = (parameter) => "#request-" + parameter.replace(/\./g, "-").toLowerCase();
var normalizeParameter = (value) => {
  if (typeof value === "string") {
    return value;
  }
  if (Array.isArray(value) && typeof value[0] === "string") {
    return value[0];
  }
  return;
};
var buildDocumentationUrl = (taskType, model, parameter, rawCode) => {
  if (model) {
    const entry = models[model];
    if (entry) {
      const base = `${DOCS_BASE}/models/${entry.id}`;
      return parameter ? `${base}${paramAnchor(parameter)}` : base;
    }
  }
  if (model) {
    const cachedBase = getDocsUrlForModel(model);
    if (cachedBase) {
      return parameter ? `${cachedBase}${paramAnchor(parameter)}` : cachedBase;
    }
  }
  if (taskType && UTILITY_DOC_PATHS[taskType]) {
    const base = `${DOCS_BASE}/${UTILITY_DOC_PATHS[taskType]}`;
    return parameter ? `${base}${paramAnchor(parameter)}` : base;
  }
  if (SDK_ONLY_RAW_CODES.has(rawCode)) {
    return `${DOCS_BASE}/${SDK_ERROR_DOC_PATH}`;
  }
  if (model) {
    return `${DOCS_BASE}/models`;
  }
  return;
};
var createRunwareError = (rawCode, message, details) => {
  const error = new RunwareError(rawCode, message);
  if (details?.parameter) {
    error.parameter = details.parameter;
  }
  if (details?.taskType) {
    error.taskType = details.taskType;
  }
  if (details?.taskUUID) {
    error.taskUUID = details.taskUUID;
  }
  if (details?.statusCode !== undefined) {
    error.statusCode = details.statusCode;
  }
  if (details?.validationErrors !== undefined && error.code === "validation") {
    error.validationErrors = details.validationErrors;
  }
  const url = buildDocumentationUrl(error.taskType, details?.model, error.parameter, rawCode);
  if (url) {
    error.documentation = url;
  }
  return error;
};
var parseApiError = (raw, context) => {
  if (typeof raw !== "object" || raw === null) {
    return createRunwareError("unknown", String(raw));
  }
  const first = Array.isArray(raw) ? raw[0] ?? {} : (() => {
    const obj = raw;
    if (Array.isArray(obj.errors)) {
      return obj.errors[0] ?? {};
    }
    if (obj.error && typeof obj.error === "object" && !Array.isArray(obj.error)) {
      return {
        ...obj.error,
        taskUUID: obj.error.taskUUID ?? obj.taskUUID,
        taskType: obj.error.taskType ?? obj.taskType
      };
    }
    return obj;
  })();
  return createRunwareError(first.code ?? "unknown", first.message ?? "An unknown API error occurred", {
    parameter: normalizeParameter(first.parameter),
    taskType: first.taskType ?? context?.taskType,
    taskUUID: first.taskUUID,
    model: first.model ?? context?.model
  });
};
var isRunwareError = (error) => error instanceof RunwareError;

// src/_version.ts
var SDK_VERSION = "1.6.7";

// src/_schemas-version.ts
var SCHEMAS_VERSION = "20260709212957";

// src/user-agent.ts
var base;
var baseUserAgent = () => {
  if (base !== undefined) {
    return base;
  }
  const runtime = typeof process !== "undefined" && process.versions?.node ? `node/${process.versions.node}; ${process.platform} ${process.arch}` : "browser";
  base = `runware-typescript/${SDK_VERSION} (${runtime}) schemas/${SCHEMAS_VERSION}`;
  return base;
};
var userAgent = (prefix) => prefix ? `${prefix} ${baseUserAgent()}` : baseUserAgent();

// src/stream.ts
var parseSseLine = (line) => {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith(":")) {
    return null;
  }
  if (!trimmed.startsWith("data:")) {
    return null;
  }
  const payload = trimmed.slice(5).trim();
  if (payload === "[DONE]") {
    return null;
  }
  let json;
  try {
    json = JSON.parse(payload);
  } catch {
    throw createRunwareError("parseError", `Failed to parse SSE data: ${payload.slice(0, 200)}`);
  }
  if (json.errors) {
    throw parseApiError(json);
  }
  return {
    text: json.delta?.text ?? undefined,
    reasoningContent: json.delta?.reasoningContent ?? undefined,
    finishReason: json.finishReason ?? null,
    usage: json.usage ?? undefined,
    cost: json.cost ?? undefined,
    resultIndex: json.resultIndex ?? undefined,
    taskUUID: json.taskUUID
  };
};
var createChunkIterator = (tasks, config, signal) => {
  const fetchImpl = config.dependencies?.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    throw createRunwareError("noFetchImpl", "fetch is required for SSE streaming");
  }
  const url = config.httpBaseUrl;
  return {
    [Symbol.asyncIterator]() {
      let reader = null;
      let decoder = null;
      let buffer = "";
      let done = false;
      const controller = new AbortController;
      let abortListener = null;
      const cleanupSignal = () => {
        if (signal && abortListener) {
          signal.removeEventListener("abort", abortListener);
          abortListener = null;
        }
      };
      const init = async () => {
        if (signal) {
          if (signal.aborted) {
            controller.abort();
          } else {
            abortListener = () => controller.abort();
            signal.addEventListener("abort", abortListener, { once: true });
          }
        }
        const body = JSON.stringify(tasks);
        config.log.send(body);
        const response = await fetchImpl(url, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${config.apiKey}`,
            "Content-Type": "application/json",
            "User-Agent": userAgent(config.userAgentPrefix)
          },
          body,
          signal: controller.signal
        });
        if (!response.ok) {
          let body2;
          try {
            body2 = await response.json();
          } catch {
            body2 = null;
          }
          if (body2) {
            throw parseApiError(body2);
          }
          throw createRunwareError("httpError", `SSE request failed: ${response.status} ${response.statusText}`);
        }
        if (!response.body) {
          throw createRunwareError("streamUnsupported", "Response body is not readable (streaming not supported)");
        }
        reader = response.body.getReader();
        decoder = new TextDecoder;
      };
      let initPromise = null;
      const iterator = {
        [Symbol.asyncIterator]() {
          return iterator;
        },
        async next() {
          if (done) {
            return { done: true, value: undefined };
          }
          if (!initPromise) {
            initPromise = init();
          }
          await initPromise;
          while (true) {
            const newlineIdx = buffer.indexOf(`
`);
            if (newlineIdx !== -1) {
              const line = buffer.slice(0, newlineIdx);
              buffer = buffer.slice(newlineIdx + 1);
              try {
                const chunk = parseSseLine(line);
                if (chunk) {
                  return { done: false, value: chunk };
                }
                if (line.trim() === "data: [DONE]") {
                  done = true;
                  reader?.cancel();
                  cleanupSignal();
                  return { done: true, value: undefined };
                }
                continue;
              } catch (error) {
                done = true;
                reader?.cancel();
                cleanupSignal();
                throw error;
              }
            }
            if (!reader || !decoder) {
              done = true;
              cleanupSignal();
              return { done: true, value: undefined };
            }
            const { done: streamDone, value } = await reader.read();
            if (streamDone) {
              done = true;
              cleanupSignal();
              return { done: true, value: undefined };
            }
            buffer += decoder.decode(value, { stream: true });
          }
        },
        async return() {
          done = true;
          reader?.cancel();
          cleanupSignal();
          return { done: true, value: undefined };
        }
      };
      return iterator;
    }
  };
};
var createTextStream = (tasks, config, signal, onStart) => {
  const chunks = [];
  let streamDone = false;
  let streamError = null;
  let consumeStarted = false;
  const waiters = new Set;
  const notifyAll = () => {
    for (const resolve of waiters) {
      resolve();
    }
    waiters.clear();
  };
  const waitForChunk = async () => new Promise((resolve) => {
    waiters.add(resolve);
  });
  const startConsuming = async () => {
    if (consumeStarted) {
      return;
    }
    consumeStarted = true;
    try {
      if (onStart) {
        await onStart();
      }
      for await (const chunk of createChunkIterator(tasks, config, signal)) {
        config.log.receive(JSON.stringify(chunk));
        chunks.push(chunk);
        notifyAll();
      }
    } catch (error) {
      streamError = error;
    } finally {
      streamDone = true;
      notifyAll();
    }
  };
  const createConsumer = (extract) => ({
    async* [Symbol.asyncIterator]() {
      startConsuming();
      let cursor = 0;
      while (true) {
        while (cursor < chunks.length) {
          const chunk = chunks[cursor];
          cursor += 1;
          if (!chunk) {
            continue;
          }
          const value = extract(chunk);
          if (value !== undefined) {
            yield value;
          }
        }
        if (streamDone) {
          if (streamError) {
            throw streamError;
          }
          return;
        }
        await waitForChunk();
      }
    }
  });
  let resultPromise = null;
  const getResult = async () => {
    if (resultPromise) {
      return resultPromise;
    }
    resultPromise = (async () => {
      startConsuming();
      while (!streamDone) {
        await waitForChunk();
      }
      if (streamError) {
        throw streamError;
      }
      let text = "";
      let reasoningContent = "";
      let finishReason = null;
      let usage;
      let cost;
      let taskUUID = "";
      for (const chunk of chunks) {
        if (chunk.text) {
          text += chunk.text;
        }
        if (chunk.reasoningContent) {
          reasoningContent += chunk.reasoningContent;
        }
        if (chunk.finishReason) {
          finishReason = chunk.finishReason;
        }
        if (chunk.usage) {
          usage = chunk.usage;
        }
        if (chunk.cost !== undefined) {
          cost = chunk.cost;
        }
        if (chunk.taskUUID) {
          taskUUID = chunk.taskUUID;
        }
      }
      return {
        text,
        reasoningContent: reasoningContent || undefined,
        finishReason,
        usage,
        cost,
        taskUUID
      };
    })();
    return resultPromise;
  };
  return {
    text: async () => (await getResult()).text,
    result: getResult,
    textStream: createConsumer((chunk) => chunk.text),
    reasoningStream: createConsumer((chunk) => chunk.reasoningContent)
  };
};

// src/logger.ts
var categories = [
  "connection",
  "auth",
  "heartbeat",
  "send",
  "receive",
  "request",
  "retry",
  "error",
  "warn",
  "info"
];
var formatData = (data) => {
  if (data === undefined || data === null) {
    return "";
  }
  if (typeof data === "string") {
    return data;
  }
  try {
    return JSON.stringify(data, null, 2);
  } catch {
    return "[unserializable]";
  }
};
var defaultSink = (entry) => {
  const line = `[RUNWARE] [${entry.category.toUpperCase()}] ${entry.timestamp} ${entry.message}`;
  const output = entry.data !== undefined ? `${line}
${formatData(entry.data)}` : line;
  if (entry.category === "error") {
    console.error(output);
  } else if (entry.category === "warn") {
    console.warn(output);
  } else {
    console.log(output);
  }
};
var noop = () => {};
var noopLogger = {
  connection: noop,
  auth: noop,
  heartbeat: noop,
  send: noop,
  receive: noop,
  request: noop,
  retry: noop,
  error: noop,
  warn: noop,
  info: noop
};
var createLogger = (enabled, sink = defaultSink) => {
  if (!enabled) {
    return noopLogger;
  }
  return Object.fromEntries(categories.map((category) => [
    category,
    (message, data) => sink({
      category,
      message,
      data,
      timestamp: new Date().toISOString()
    })
  ]));
};

// src/config.ts
var DEFAULT_CONFIG = {
  transport: "websocket",
  wsBaseUrl: "wss://ws-api.runware.ai/v1",
  httpBaseUrl: "https://api.runware.ai/v1",
  timeout: 1200000,
  pollTimeout: 1200000,
  authTimeout: 15000,
  maxRetries: 3,
  retryDelay: 1000,
  retryStrategy: "exponential",
  maxReconnectAttempts: Infinity,
  debug: false
};
var createConfig = (config) => {
  if (!config.apiKey) {
    throw createRunwareError("missingApiKey", "API key is required");
  }
  const merged = { ...DEFAULT_CONFIG, ...config };
  merged.log = createLogger(merged.debug ?? false, merged.logSink);
  return merged;
};
var createBrowserDependencies = () => ({
  WebSocket: globalThis.WebSocket,
  fetch: globalThis.fetch
});
var createNodeDependencies = async () => {
  const nodeDeps = { fetch: globalThis.fetch };
  try {
    const wsModule = await import("ws");
    nodeDeps.WebSocket = wsModule.default;
  } catch {
    if (globalThis.WebSocket) {
      nodeDeps.WebSocket = globalThis.WebSocket;
    }
  }
  return nodeDeps;
};
var isNodeJS = () => typeof process !== "undefined" && typeof process.versions?.node === "string";

// src/transport/websocket.ts
var RECONNECT_DELAY_CAP_MS = 30000;
var RECONNECT_JITTER_MS = 1000;
var computeReconnectDelay = (retryDelay, attempt, randomFn = Math.random) => {
  const base2 = retryDelay * Math.pow(2, attempt - 1);
  const capped = Math.min(base2, RECONNECT_DELAY_CAP_MS);
  const jitter = randomFn() * RECONNECT_JITTER_MS;
  return capped + jitter;
};
var createWebSocketTransport = (config) => {
  let ws = null;
  let connectionState = {
    connected: false,
    lastActivity: Date.now()
  };
  let pingInterval = null;
  let isReconnecting = false;
  let reconnectAttempt = 0;
  let shouldReconnect = true;
  let reconnectTimer = null;
  const taskCallbacks = new Map;
  const setupHeartbeat = () => {
    if (pingInterval) {
      clearInterval(pingInterval);
    }
    pingInterval = setInterval(() => {
      if (!connectionState.connected) {
        return;
      }
      try {
        const pingMessage = JSON.stringify([{ taskType: "ping", ping: true }]);
        ws.send(pingMessage);
      } catch (error) {
        config.log.heartbeat("Ping failed, reconnecting", error);
        reconnect();
      }
      const inactivityTime = Date.now() - connectionState.lastActivity;
      if (inactivityTime > 1e5) {
        config.log.heartbeat(`No activity for ${inactivityTime}ms, reconnecting`);
        reconnect();
      }
    }, 30000);
  };
  const safeCall = (cb, frame) => {
    try {
      cb(frame);
    } catch (error) {
      config.log.error("task callback raised in dispatch", error);
    }
  };
  const handleMessage = (event) => {
    connectionState.lastActivity = Date.now();
    try {
      const dataString = typeof event.data === "string" ? event.data : event.data.toString();
      config.log.receive(dataString);
      const response = JSON.parse(dataString);
      const rawErrors = response.errors ?? response.error;
      if (rawErrors) {
        const errors = Array.isArray(rawErrors) ? rawErrors : [rawErrors];
        const errorsByTask = new Map;
        const unrouted = [];
        for (const error of errors) {
          const taskUUID = error.taskUUID;
          if (taskUUID) {
            const list = errorsByTask.get(taskUUID) ?? [];
            list.push(error);
            errorsByTask.set(taskUUID, list);
          } else {
            unrouted.push(error);
          }
        }
        let delivered = false;
        for (const [taskUUID, taskErrors] of errorsByTask) {
          const callback = taskCallbacks.get(taskUUID);
          if (callback) {
            delivered = true;
            safeCall(callback, { error: taskErrors });
          }
        }
        if (unrouted.length > 0 || !delivered) {
          config.log.error("Unroutable WebSocket error", errors);
          for (const [, callback] of taskCallbacks) {
            safeCall(callback, { error: errors });
          }
        }
        return;
      }
      if (!response.data || !Array.isArray(response.data) || response.data.length === 0) {
        config.log.warn("Received message without valid data array", response);
        return;
      }
      const itemsByTask = new Map;
      for (const item of response.data) {
        if (item.taskType === "ping" && item.pong === true) {
          continue;
        }
        if (item.taskType === "authentication") {
          if (item.connectionSessionUUID) {
            connectionState.connectionSessionUUID = item.connectionSessionUUID;
          }
          continue;
        }
        const taskUUID = item.taskUUID;
        if (!taskUUID) {
          continue;
        }
        const list = itemsByTask.get(taskUUID) ?? [];
        list.push(item);
        itemsByTask.set(taskUUID, list);
      }
      for (const [taskUUID, items] of itemsByTask) {
        const callback = taskCallbacks.get(taskUUID);
        if (callback) {
          safeCall(callback, { data: items });
        } else {
          config.log.warn(`No callback for taskUUID: ${taskUUID}`);
        }
      }
    } catch (error) {
      config.log.error("Failed to parse WebSocket message", error);
    }
  };
  const establishConnection = async () => {
    if (!config.dependencies?.WebSocket) {
      throw createRunwareError("noWebSocketImpl", "WebSocket implementation is required");
    }
    shouldReconnect = true;
    const WebSocketImpl = config.dependencies.WebSocket;
    if (ws) {
      const stale = ws;
      ws = null;
      connectionState.connected = false;
      if (pingInterval) {
        clearInterval(pingInterval);
        pingInterval = null;
      }
      try {
        if (stale.readyState === 0 || stale.readyState === 1) {
          stale.close();
        }
      } catch {}
    }
    return new Promise((resolve, reject) => {
      let settled = false;
      try {
        const wsOptions = { headers: { "User-Agent": userAgent(config.userAgentPrefix) } };
        ws = new WebSocketImpl(config.wsBaseUrl, [], wsOptions);
        const thisSocket = ws;
        ws.onopen = async () => {
          if (ws !== thisSocket) {
            return;
          }
          connectionState.connected = true;
          try {
            await authenticate();
            setupHeartbeat();
            if (!settled) {
              settled = true;
              resolve();
            }
          } catch (error) {
            ws?.close();
            if (!settled) {
              settled = true;
              reject(error);
            }
          }
        };
        ws.onmessage = handleMessage;
        ws.onclose = () => {
          if (ws !== thisSocket) {
            return;
          }
          connectionState.connected = false;
          if (pingInterval) {
            clearInterval(pingInterval);
          }
          pingInterval = null;
          ws = null;
          config.log.connection("Connection closed");
          if (shouldReconnect && !isReconnecting) {
            reconnect();
          }
        };
        ws.onerror = (errorEvent) => {
          if (ws !== thisSocket) {
            return;
          }
          const error = errorEvent.error ?? errorEvent;
          config.log.error("WebSocket error", error);
          if (!settled && !connectionState.connected) {
            settled = true;
            reject(createRunwareError("connectionFailed", `WebSocket connection failed: ${error?.message ?? "Unknown error"}`));
            return;
          }
          ws?.close();
          reconnect();
        };
      } catch (error) {
        if (!settled) {
          settled = true;
          reject(error);
        }
      }
    });
  };
  const authenticate = async () => {
    return new Promise((resolve, reject) => {
      if (!ws || ws.readyState !== 1) {
        return reject(createRunwareError("notOpen", "WebSocket not open for authentication"));
      }
      const authMessage = JSON.stringify([{
        taskType: "authentication",
        apiKey: config.apiKey,
        ...connectionState.connectionSessionUUID ? { connectionSessionUUID: connectionState.connectionSessionUUID } : {}
      }]);
      let authTimeoutId = null;
      const originalMessageHandler = ws.onmessage;
      const restoreHandler = () => {
        if (ws) {
          ws.onmessage = originalMessageHandler;
        }
      };
      const authHandler = (event) => {
        try {
          const dataString = typeof event.data === "string" ? event.data : event.data.toString();
          const response = JSON.parse(dataString);
          if (response.data?.[0]?.taskType === "authentication") {
            if (authTimeoutId) {
              clearTimeout(authTimeoutId);
            }
            restoreHandler();
            if (response.data[0].connectionSessionUUID) {
              connectionState.connectionSessionUUID = response.data[0].connectionSessionUUID;
              resolve();
            } else {
              reject(createRunwareError("authFailed", "Authentication successful but missing connectionSessionUUID"));
            }
            return;
          }
          if (response.error) {
            if (authTimeoutId) {
              clearTimeout(authTimeoutId);
            }
            restoreHandler();
            reject(createRunwareError("authFailed", response.error.message ?? "Authentication failed via API response"));
            return;
          }
          if (originalMessageHandler) {
            originalMessageHandler(event);
          }
        } catch (parseError) {
          if (originalMessageHandler) {
            originalMessageHandler(event);
          }
        }
      };
      ws.onmessage = authHandler;
      authTimeoutId = setTimeout(() => {
        restoreHandler();
        reject(createRunwareError("authTimeout", `Authentication timed out after ${config.authTimeout}ms`));
      }, config.authTimeout);
      try {
        config.log.send(JSON.stringify([{
          taskType: "authentication",
          apiKey: "[redacted]",
          ...connectionState.connectionSessionUUID ? { connectionSessionUUID: connectionState.connectionSessionUUID } : {}
        }]));
        ws.send(authMessage);
      } catch (sendError) {
        if (authTimeoutId) {
          clearTimeout(authTimeoutId);
        }
        restoreHandler();
        reject(createRunwareError("sendFailed", `Failed to send authentication message: ${sendError}`));
      }
    });
  };
  const reconnect = async () => {
    if (!shouldReconnect) {
      return;
    }
    if (isReconnecting) {
      return;
    }
    isReconnecting = true;
    if (ws) {
      try {
        connectionState.connected = false;
        if (pingInterval) {
          clearInterval(pingInterval);
          pingInterval = null;
        }
        if (ws.readyState === 0 || ws.readyState === 1) {
          ws.close();
        }
        ws = null;
      } catch {}
    }
    try {
      await establishConnection();
      if (!shouldReconnect) {
        isReconnecting = false;
        if (ws) {
          try {
            ws.close();
          } catch {}
          ws = null;
        }
        return;
      }
      config.log.connection(`Reconnected with ${taskCallbacks.size} pending task(s)`);
      reconnectAttempt = 0;
      isReconnecting = false;
    } catch (error) {
      reconnectAttempt += 1;
      if (!shouldReconnect) {
        isReconnecting = false;
        return;
      }
      if (reconnectAttempt >= config.maxReconnectAttempts) {
        config.log.error(`Reconnection failed after ${reconnectAttempt} attempts, giving up`);
        isReconnecting = false;
        const reconnectError = createRunwareError("reconnectionFailed", `Permanently disconnected after ${reconnectAttempt} reconnection attempts`);
        const errorList = [reconnectError];
        const errorFrame = { error: errorList };
        for (const [, callback] of taskCallbacks) {
          safeCall(callback, errorFrame);
        }
        taskCallbacks.clear();
        return;
      }
      const delay = computeReconnectDelay(config.retryDelay, reconnectAttempt);
      const max = config.maxReconnectAttempts === Infinity ? "∞" : config.maxReconnectAttempts;
      config.log.error(`Reconnection failed (attempt ${reconnectAttempt}/${max}),` + ` retrying in ${Math.round(delay)}ms`, error);
      reconnectTimer = setTimeout(() => {
        reconnectTimer = null;
        isReconnecting = false;
        reconnect();
      }, delay);
    }
  };
  const connect = async () => {
    reconnectAttempt = 0;
    return establishConnection();
  };
  const disconnect = async () => {
    shouldReconnect = false;
    if (reconnectTimer) {
      clearTimeout(reconnectTimer);
      reconnectTimer = null;
    }
    if (!ws) {
      return Promise.resolve();
    }
    return new Promise((resolve) => {
      connectionState.connected = false;
      if (pingInterval) {
        clearInterval(pingInterval);
        pingInterval = null;
      }
      const disconnectError = { error: [{ code: "disconnected", message: "Client disconnected" }] };
      for (const [, callback] of taskCallbacks) {
        safeCall(callback, disconnectError);
      }
      taskCallbacks.clear();
      if (ws.readyState === 0 || ws.readyState === 1) {
        ws.onclose = () => {
          ws = null;
          resolve();
        };
        ws.onerror = () => {
          ws = null;
          resolve();
        };
        ws.close();
      } else {
        ws = null;
        resolve();
      }
    });
  };
  const redactApiKey = (payload) => payload.map((task) => task.taskType === "authentication" && ("apiKey" in task) ? { ...task, apiKey: "[redacted]" } : task);
  const sendRequest = async (data) => {
    if (!connectionState.connected) {
      throw createRunwareError("notConnected", "Not connected to WebSocket server");
    }
    try {
      const payload = Array.isArray(data) ? data : [data];
      const serialized = JSON.stringify(payload);
      config.log.send(JSON.stringify(redactApiKey(payload)));
      ws.send(serialized);
      return Promise.resolve();
    } catch (error) {
      return Promise.reject(error);
    }
  };
  const subscribeToTask = (taskUUID, callback) => {
    taskCallbacks.set(taskUUID, callback);
  };
  const unsubscribeFromTask = (taskUUID) => {
    taskCallbacks.delete(taskUUID);
  };
  const isConnected = () => {
    return connectionState.connected;
  };
  const getSessionId = () => {
    return connectionState.connectionSessionUUID;
  };
  return {
    type: "websocket",
    connect,
    disconnect,
    isConnected,
    getSessionId,
    sendRequest,
    subscribeToTask,
    unsubscribeFromTask
  };
};

// src/utils/retry.ts
var calculateRetryDelay = (attempt, baseDelay, strategy) => {
  let delay;
  if (strategy === "linear") {
    delay = baseDelay * (attempt + 1);
  } else {
    const fullDelay = baseDelay * Math.pow(2, attempt);
    const jitter = Math.random() * 0.3 * fullDelay;
    delay = fullDelay + jitter;
  }
  return Math.min(delay, 30000);
};
var interruptibleSleep = async (ms, signal) => new Promise((resolve) => {
  if (!signal) {
    setTimeout(resolve, ms);
    return;
  }
  if (signal.aborted) {
    resolve();
    return;
  }
  const timer = setTimeout(() => {
    signal.removeEventListener("abort", onAbort);
    resolve();
  }, ms);
  const onAbort = () => {
    clearTimeout(timer);
    resolve();
  };
  signal.addEventListener("abort", onAbort, { once: true });
});
var withRetry = async (fn, options) => {
  const {
    maxRetries,
    retryDelay,
    retryStrategy = "exponential",
    shouldRetry = () => false,
    onRetry,
    signal
  } = options;
  let lastError;
  for (let attempt = 0;attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;
      if (attempt === maxRetries || !shouldRetry(error)) {
        break;
      }
      const delay = calculateRetryDelay(attempt, retryDelay, retryStrategy);
      if (onRetry) {
        onRetry(error, attempt + 1, delay);
      }
      await interruptibleSleep(delay, signal);
      if (signal?.aborted) {
        break;
      }
    }
  }
  throw lastError;
};

// src/transport/rest.ts
var createTimeoutError = (message) => {
  const error = new Error(message);
  error.name = "TimeoutError";
  error.isTimeoutError = true;
  return error;
};
var isTimeoutError = (error) => error instanceof Error && error.isTimeoutError === true;
var isAbortError = (error) => error instanceof Error && error.name === "AbortError";
var isRetryableStatus = (status) => status === 408 || status === 429 || status >= 500 && status < 600;
var createRestTransport = (config) => {
  const customFetch = async (url, options) => {
    const fetchImpl = config.dependencies?.fetch ?? globalThis.fetch;
    if (!fetchImpl) {
      throw createRunwareError("noFetchImpl", "Fetch implementation is required for REST transport");
    }
    return fetchImpl(url, options);
  };
  const sendRequest = async (data, options) => {
    const payload = Array.isArray(data) ? data : [data];
    const url = config.httpBaseUrl;
    const requestTimeout = options?.timeout ?? config.timeout;
    try {
      const response = await fetchWithRetry(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${config.apiKey}`,
          "User-Agent": userAgent(config.userAgentPrefix)
        },
        body: JSON.stringify(payload)
      }, requestTimeout, options?.signal);
      return response;
    } catch (error) {
      if (isTimeoutError(error)) {
        throw createRunwareError("timeout", `REST request timed out after ${requestTimeout}ms (gave up after ${config.maxRetries + 1} attempts)`);
      }
      throw error;
    }
  };
  const fetchWithRetry = async (url, options, requestTimeout, externalSignal) => {
    return withRetry(async () => {
      const controller = new AbortController;
      const timeoutId = setTimeout(() => controller.abort(), requestTimeout);
      const onExternalAbort = () => controller.abort();
      if (externalSignal) {
        if (externalSignal.aborted) {
          controller.abort();
        } else {
          externalSignal.addEventListener("abort", onExternalAbort, { once: true });
        }
      }
      try {
        const response = await customFetch(url, {
          ...options,
          signal: controller.signal
        });
        if (!response.ok) {
          let body;
          try {
            body = await response.json();
          } catch {
            body = null;
          }
          if (body) {
            let context;
            try {
              const sent = JSON.parse(options.body);
              const firstTask = Array.isArray(sent) ? sent[0] : sent;
              if (firstTask) {
                context = {
                  taskType: firstTask.taskType,
                  model: firstTask.model
                };
              }
            } catch {}
            const parsed = parseApiError(body, context);
            parsed.statusCode = response.status;
            throw parsed;
          }
          if (response.status === 401) {
            throw createRunwareError("invalidApiKey", `Authentication failed: invalid API key (HTTP 401 ${response.statusText})`, { statusCode: 401 });
          }
          throw createRunwareError("httpError", `HTTP ${response.status}: ${response.statusText}`, { statusCode: response.status });
        }
        try {
          return await response.json();
        } catch (parseError) {
          throw createRunwareError("parseError", `Failed to parse JSON response (HTTP ${response.status}): ${parseError instanceof Error ? parseError.message : String(parseError)}`, { statusCode: response.status });
        }
      } catch (error) {
        if (isAbortError(error)) {
          if (externalSignal?.aborted) {
            throw createRunwareError("aborted", "Request aborted");
          }
          throw createTimeoutError(`Request timed out after ${requestTimeout}ms`);
        }
        throw error;
      } finally {
        clearTimeout(timeoutId);
        if (externalSignal) {
          externalSignal.removeEventListener("abort", onExternalAbort);
        }
      }
    }, {
      maxRetries: config.maxRetries,
      retryDelay: config.retryDelay,
      retryStrategy: config.retryStrategy,
      signal: externalSignal,
      shouldRetry: (error) => {
        if (isRunwareError(error) && error.code === "aborted") {
          return false;
        }
        if (isTimeoutError(error)) {
          return true;
        }
        if (isRunwareError(error) && error.statusCode !== undefined) {
          if (isRetryableStatus(error.statusCode)) {
            return true;
          }
        }
        if (error instanceof TypeError) {
          return true;
        }
        return false;
      },
      onRetry: (error, attempt, delayMs) => {
        const reason = (() => {
          if (isTimeoutError(error)) {
            return "timeout";
          }
          if (isRunwareError(error) && error.statusCode) {
            return `HTTP ${error.statusCode}`;
          }
          if (error instanceof TypeError) {
            return "network error";
          }
          return "unknown";
        })();
        config.log.retry(`Retrying after ${reason}, attempt ${attempt}/${config.maxRetries} in ${Math.round(delayMs)}ms`);
      }
    });
  };
  return { type: "rest", sendRequest };
};

// src/transport/index.ts
var createTransport = (type, config) => {
  if (type === "websocket") {
    return createWebSocketTransport(config);
  } else if (type === "rest") {
    return createRestTransport(config);
  } else {
    throw new Error(`Unsupported transport type: ${type}`);
  }
};

// src/registry.ts
var DEFAULT_TTL = 300000;
var DEFAULT_INITIAL_FETCH_TIMEOUT = 3000;
var createRegistry = (options) => {
  const fetchImpl = options.fetchImpl ?? globalThis.fetch;
  const ttl = options.ttl ?? DEFAULT_TTL;
  const initialFetchTimeout = options.initialFetchTimeout ?? DEFAULT_INITIAL_FETCH_TIMEOUT;
  let cached = null;
  let etag;
  let lastFetchAt = 0;
  let inFlightFetch = null;
  let initialFetchAttempted = false;
  let slugToAir = null;
  let slugIndexSource = null;
  const getSlugIndex = () => {
    if (slugToAir && slugIndexSource === cached) {
      return slugToAir;
    }
    const index = new Map;
    for (const [air, entry] of Object.entries(options.fallback.models)) {
      if (entry.id) {
        index.set(entry.id, air);
      }
    }
    if (cached) {
      for (const [air, entry] of Object.entries(cached.models)) {
        if (entry.id) {
          index.set(entry.id, air);
        }
      }
    }
    slugToAir = index;
    slugIndexSource = cached;
    return index;
  };
  const fetchRegistry = async () => {
    if (!fetchImpl) {
      return;
    }
    if (inFlightFetch) {
      return inFlightFetch;
    }
    inFlightFetch = (async () => {
      try {
        const headers = {};
        if (etag) {
          headers["If-None-Match"] = etag;
        }
        const response = await fetchImpl(options.url, { headers });
        if (response.status === 304) {
          options.log.info("Registry not modified (304)");
          lastFetchAt = Date.now();
          return;
        }
        if (!response.ok) {
          throw new Error(`Registry fetch failed: HTTP ${response.status}`);
        }
        const data = await response.json();
        cached = data;
        etag = response.headers.get("etag") ?? undefined;
        lastFetchAt = Date.now();
        options.log.info(`Registry refreshed (version ${data.version})`);
      } finally {
        inFlightFetch = null;
      }
    })();
    return inFlightFetch;
  };
  const refreshSilently = () => {
    fetchRegistry().catch((error) => {
      options.log.warn("Background registry refresh failed", error);
    });
  };
  const ensureInitialized = async () => {
    if (initialFetchAttempted) {
      return;
    }
    initialFetchAttempted = true;
    const fetchPromise = fetchRegistry().catch((error) => {
      options.log.warn("Initial registry fetch failed, using fallback", error);
    });
    let timeoutId = null;
    const timeoutPromise = new Promise((resolve) => {
      timeoutId = setTimeout(() => {
        options.log.warn(`Registry fetch timed out after ${initialFetchTimeout}ms, using fallback`);
        resolve();
      }, initialFetchTimeout);
    });
    await Promise.race([fetchPromise, timeoutPromise]);
    if (timeoutId) {
      clearTimeout(timeoutId);
    }
  };
  const ensureFresh = async () => {
    await ensureInitialized();
    if (!cached) {
      return;
    }
    if (Date.now() - lastFetchAt < ttl) {
      return;
    }
    refreshSilently();
  };
  const lookup = async (key, field) => {
    await ensureFresh();
    if (cached?.[field][key]) {
      return cached[field][key];
    }
    if (options.fallback[field][key]) {
      return options.fallback[field][key];
    }
    return null;
  };
  const resolveModelAir = async (input) => {
    await ensureFresh();
    if (cached?.models[input] || options.fallback.models[input]) {
      return input;
    }
    return getSlugIndex().get(input) ?? null;
  };
  const lookupModel = async (model) => {
    const air = await resolveModelAir(model);
    if (!air) {
      return null;
    }
    return cached?.models[air] ?? options.fallback.models[air] ?? null;
  };
  const getModelTaskType = async (model) => (await lookupModel(model))?.taskType ?? null;
  const getModelId = async (model) => (await lookupModel(model))?.id ?? null;
  const getArchitectureTaskType = async (key) => lookup(key, "architectureTaskTypes");
  const getModalityTaskType = async (key) => lookup(key, "modalityTaskTypes");
  const refresh = async () => {
    initialFetchAttempted = true;
    await fetchRegistry();
  };
  const notifyVersion = (version) => {
    if (cached?.version === version) {
      return;
    }
    options.log.info(`Observed new registry version ${version}, refreshing`);
    refreshSilently();
  };
  const getVersion = () => cached?.version;
  return {
    getModelTaskType,
    getModelId,
    resolveModelAir,
    getArchitectureTaskType,
    getModalityTaskType,
    refresh,
    notifyVersion,
    getVersion
  };
};

// src/content.ts
var buildQuery = (params) => {
  const entries = [];
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null) {
      continue;
    }
    entries.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`);
  }
  return entries.length ? `?${entries.join("&")}` : "";
};
var fetchJson = async (url, fetchImpl, treat404AsNull) => {
  let response;
  try {
    response = await fetchImpl(url);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw createRunwareError("connectionFailed", `Failed to reach content service: ${message}`);
  }
  if (response.status === 404 && treat404AsNull) {
    return null;
  }
  if (!response.ok) {
    throw createRunwareError("httpError", `Content service responded ${response.status} for ${url}`);
  }
  return await response.json();
};
var NON_NATIVE_STATUSES = new Set(["openai-compatible"]);
var excludeNonNative = (models2) => models2.filter((model) => !NON_NATIVE_STATUSES.has(model.status ?? ""));
var createContentClient = (config) => {
  const fetchImpl = config.dependencies?.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    throw createRunwareError("missingDependency", "No fetch implementation available. Provide config.dependencies.fetch.");
  }
  const baseUrl = CONTENT_BASE_URL;
  const listModels = async (opts = {}) => {
    const query = buildQuery({
      capability: opts.capability,
      status: opts.status,
      category: opts.category,
      creator: opts.creator,
      q: opts.search,
      sort: opts.sort,
      order: opts.order,
      paginate: opts.paginate ? "true" : undefined,
      limit: opts.limit,
      offset: opts.offset
    });
    const url = `${baseUrl}/models${query}`;
    const result = await fetchJson(url, fetchImpl, false);
    if (opts.status || result === null) {
      return result;
    }
    return Array.isArray(result) ? excludeNonNative(result) : { ...result, items: excludeNonNative(result.items) };
  };
  const getModel = async (modelId) => fetchJson(`${baseUrl}/models/${encodeURIComponent(modelId)}`, fetchImpl, true);
  const getModelExamples = async (modelId, opts = {}) => {
    const query = buildQuery({ capability: opts.capability });
    const url = `${baseUrl}/models/${encodeURIComponent(modelId)}/examples${query}`;
    return await fetchJson(url, fetchImpl, true) ?? [];
  };
  const getModelPricing = async (modelId) => fetchJson(`${baseUrl}/models/${encodeURIComponent(modelId)}/pricing`, fetchImpl, true);
  const getModelGuides = async (modelId) => await fetchJson(`${baseUrl}/models/${encodeURIComponent(modelId)}/guides`, fetchImpl, true) ?? [];
  const listCollections = async (opts = {}) => {
    const query = buildQuery({ category: opts.category });
    const url = `${baseUrl}/collections${query}`;
    return await fetchJson(url, fetchImpl, false) ?? [];
  };
  const getCollection = async (collectionId) => fetchJson(`${baseUrl}/collections/${encodeURIComponent(collectionId)}`, fetchImpl, true);
  const listCapabilities = async () => await fetchJson(`${baseUrl}/capabilities`, fetchImpl, false) ?? [];
  const listCreators = async () => await fetchJson(`${baseUrl}/creators`, fetchImpl, false) ?? [];
  const getCreator = async (creatorId) => fetchJson(`${baseUrl}/creators/${encodeURIComponent(creatorId)}`, fetchImpl, true);
  return {
    listModels,
    getModel,
    getModelExamples,
    getModelPricing,
    getModelGuides,
    listCollections,
    getCollection,
    listCapabilities,
    listCreators,
    getCreator
  };
};

// src/utils/file.ts
var fileToDataURI = async (file) => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader;
    reader.onload = () => {
      if (typeof reader.result === "string") {
        resolve(reader.result);
      } else {
        reject(new Error("FileReader did not return a string"));
      }
    };
    reader.onerror = () => reject(reader.error);
    reader.readAsDataURL(file);
  });
};
var MAX_PATH_LEN = 4096;
var REMOTE_PREFIXES = ["http://", "https://", "data:"];
var isNode = () => typeof process !== "undefined" && typeof process.versions?.node === "string";
var encodeLocalFiles = async (value) => {
  if (!isNode()) {
    return value;
  }
  const fs = await import("node:fs/promises");
  const encodeString = async (str) => {
    if (REMOTE_PREFIXES.some((p) => str.startsWith(p))) {
      return str;
    }
    if (str.length > MAX_PATH_LEN) {
      return str;
    }
    try {
      const stat = await fs.stat(str);
      if (!stat.isFile()) {
        return str;
      }
    } catch {
      return str;
    }
    const bytes = await fs.readFile(str);
    return bytes.toString("base64");
  };
  const walk = async (node) => {
    if (typeof node === "string") {
      return encodeString(node);
    }
    if (Array.isArray(node)) {
      return Promise.all(node.map(walk));
    }
    if (node !== null && typeof node === "object") {
      const entryPromises = Object.entries(node).map(async ([k, v]) => [k, await walk(v)]);
      const entries = await Promise.all(entryPromises);
      return Object.fromEntries(entries);
    }
    return node;
  };
  return walk(value);
};

// src/client.ts
var generateUUID = () => {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (char) => {
    const randomHex = Math.random() * 16 | 0;
    const value = char === "x" ? randomHex : randomHex & 3 | 8;
    return value.toString(16);
  });
};
var normalizeTasks = (taskType, params) => {
  const items = Array.isArray(params) ? params : [params];
  return items.map((item) => ({
    ...item,
    taskType,
    taskUUID: item.taskUUID ?? generateUUID()
  }));
};
var resolveTaskType = async (schemaKey, params, registry) => {
  if (typeof params.taskType === "string" && params.taskType.length > 0) {
    return params.taskType;
  }
  if (typeof params.model === "string") {
    const resolved = await registry.getModelTaskType(params.model);
    if (resolved) {
      return resolved;
    }
  }
  if (schemaKey) {
    const arch = await registry.getArchitectureTaskType(schemaKey);
    if (arch) {
      return arch;
    }
    const modality = await registry.getModalityTaskType(schemaKey);
    if (modality) {
      return modality;
    }
  }
  const model = typeof params.model === "string" ? params.model : "undefined";
  throw createRunwareError("unknownModel", `Unknown model '${model}'. ` + "Pass taskType in params (e.g., { taskType: 'imageInference' }). " + "The generic in .run<'image'>() provides TypeScript types but is erased at runtime. " + "If this is a newly launched model, try client.refreshRegistry() to fetch the latest models.");
};
var executeWebSocketSync = async (tasks, transport, config, options) => {
  const signal = options?.signal;
  const dispatcher = createItemDispatcher(config, options);
  const results = [];
  const expectedPerTask = new Map;
  const receivedPerTask = new Map;
  for (const task of tasks) {
    expectedPerTask.set(task.taskUUID, Number(task.numberResults) || 1);
    receivedPerTask.set(task.taskUUID, 0);
  }
  return new Promise((resolve, reject) => {
    let settled = false;
    const timeout = options?.timeout ?? config.timeout;
    const cleanup = () => {
      if (settled) {
        return;
      }
      settled = true;
      clearTimeout(timer);
      if (signal) {
        signal.removeEventListener("abort", onAbort);
      }
      for (const task of tasks) {
        transport.unsubscribeFromTask(task.taskUUID);
      }
    };
    const timer = setTimeout(() => {
      cleanup();
      reject(createRunwareError("timeout", `Task timed out after ${timeout}ms`));
    }, timeout);
    const onAbort = () => {
      cleanup();
      reject(createRunwareError("aborted", "Request aborted"));
    };
    if (signal) {
      signal.addEventListener("abort", onAbort, { once: true });
    }
    const isComplete = () => {
      for (const [uuid, expected] of expectedPerTask) {
        if ((receivedPerTask.get(uuid) ?? 0) < expected) {
          return false;
        }
      }
      return true;
    };
    for (const task of tasks) {
      transport.subscribeToTask(task.taskUUID, (response) => {
        if (settled) {
          return;
        }
        if (response.error) {
          cleanup();
          reject(parseApiError(response.error, {
            taskType: task.taskType,
            model: task.model
          }));
          return;
        }
        if (response.data) {
          results.push(...response.data);
          receivedPerTask.set(task.taskUUID, (receivedPerTask.get(task.taskUUID) ?? 0) + response.data.length);
          dispatcher.dispatch(response.data);
        }
        if (isComplete()) {
          cleanup();
          resolve(results);
        }
      });
    }
    transport.sendRequest(tasks).catch((error) => {
      cleanup();
      reject(error);
    });
  });
};
var executeWebSocketAsync = async (tasks, transport, config, options) => {
  const signal = options?.signal;
  const dispatcher = createItemDispatcher(config, options);
  const expectedPerTask = new Map;
  const resultsPerTask = new Map;
  const tasksByUUID = new Map;
  for (const task of tasks) {
    expectedPerTask.set(task.taskUUID, Number(task.numberResults) || 1);
    resultsPerTask.set(task.taskUUID, []);
    tasksByUUID.set(task.taskUUID, task);
  }
  const ackResolvers = new Map;
  const ackTimers = new Map;
  const pendingPolls = new Map;
  for (const task of tasks) {
    transport.subscribeToTask(task.taskUUID, (response) => {
      const rawErrors = response.errors ?? response.error;
      const errors = rawErrors ? Array.isArray(rawErrors) ? rawErrors : [rawErrors] : null;
      const ackResolver = ackResolvers.get(task.taskUUID);
      if (ackResolver) {
        ackResolvers.delete(task.taskUUID);
        ackResolver(errors ? { ok: false, errors } : { ok: true });
        return;
      }
      const pollResolver = pendingPolls.get(task.taskUUID);
      if (!pollResolver) {
        return;
      }
      pendingPolls.delete(task.taskUUID);
      pollResolver({ items: response.data ?? [], errors: errors ?? [] });
    });
  }
  const cleanup = () => {
    for (const task of tasks) {
      transport.unsubscribeFromTask(task.taskUUID);
    }
    for (const timer of ackTimers.values()) {
      clearTimeout(timer);
    }
    ackTimers.clear();
    ackResolvers.clear();
    pendingPolls.clear();
  };
  try {
    const ACK_TIMEOUT_MS = 30000;
    const ackPromises = tasks.map(async (task) => new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        ackTimers.delete(task.taskUUID);
        if (ackResolvers.has(task.taskUUID)) {
          ackResolvers.delete(task.taskUUID);
          reject(createRunwareError("timeout", `ACK not received for task ${task.taskUUID} after ${ACK_TIMEOUT_MS}ms`));
        }
      }, ACK_TIMEOUT_MS);
      ackTimers.set(task.taskUUID, timer);
      ackResolvers.set(task.taskUUID, (result) => {
        const timer2 = ackTimers.get(task.taskUUID);
        if (timer2) {
          clearTimeout(timer2);
          ackTimers.delete(task.taskUUID);
        }
        resolve(result);
      });
    }));
    await transport.sendRequest(tasks);
    const acks = await Promise.all(ackPromises);
    for (let i = 0;i < tasks.length; i++) {
      const ack = acks[i];
      const task = tasks[i];
      if (!ack || !task) {
        continue;
      }
      if (!ack.ok) {
        const parsed = parseApiError({ errors: ack.errors }, {
          taskType: task.taskType,
          model: task.model
        });
        if (!parsed.taskUUID) {
          parsed.taskUUID = task.taskUUID;
        }
        throw parsed;
      }
    }
    const pollTimeout = options?.timeout ?? config.pollTimeout;
    const startTime = Date.now();
    let delay = 1000;
    let pollNumber = 0;
    const pending = new Set(tasks.map((task) => task.taskUUID));
    config.log.request(`Polling ${pending.size} async task(s) over WebSocket`);
    while (pending.size > 0) {
      if (signal?.aborted) {
        throw createRunwareError("aborted", "Request aborted during polling");
      }
      if (Date.now() - startTime >= pollTimeout) {
        throw createRunwareError("timeout", `Polling timed out after ${pollTimeout}ms with ${pending.size} task(s) still processing`);
      }
      await interruptibleSleep2(delay, signal);
      if (signal?.aborted) {
        throw createRunwareError("aborted", "Request aborted during polling");
      }
      const uuids = Array.from(pending);
      const elapsed = Date.now() - startTime;
      const perPollTimeout = Math.max(1, Math.min(pollTimeout - elapsed, config.timeout));
      const polled = await Promise.all(uuids.map(async (uuid) => {
        const promise = new Promise((resolve) => {
          pendingPolls.set(uuid, resolve);
        });
        await transport.sendRequest([{
          taskType: "getResponse",
          taskUUID: uuid
        }]);
        let timerId;
        try {
          const result = await Promise.race([
            promise,
            new Promise((_, reject) => {
              timerId = setTimeout(() => reject(createRunwareError("timeout", `getResponse poll for task ${uuid} timed out after ${perPollTimeout}ms`)), perPollTimeout);
            })
          ]);
          return { uuid, result };
        } finally {
          if (timerId) {
            clearTimeout(timerId);
          }
          pendingPolls.delete(uuid);
        }
      }));
      for (const { uuid, result } of polled) {
        dispatcher.dispatch(result.items);
        const itemErrors = result.items.filter((item) => item.status === "error").map((item) => ({ ...item.error ?? {}, taskUUID: uuid }));
        const allErrors = [...result.errors, ...itemErrors];
        if (allErrors.length > 0) {
          const origTask = tasksByUUID.get(uuid);
          const parsed = parseApiError({ errors: allErrors }, {
            taskType: origTask?.taskType,
            model: origTask ? origTask.model : undefined
          });
          if (!parsed.taskUUID) {
            parsed.taskUUID = uuid;
          }
          throw parsed;
        }
        const successItems = result.items.filter((item) => item.status === "success");
        resultsPerTask.set(uuid, successItems);
        const expected = expectedPerTask.get(uuid) ?? 1;
        if (successItems.length >= expected) {
          pending.delete(uuid);
        }
      }
      pollNumber += 1;
      delay = pollNumber === 1 ? 500 : Math.min(delay * 1.5, 1e4);
    }
  } finally {
    cleanup();
  }
  const results = [];
  for (const task of tasks) {
    results.push(...resultsPerTask.get(task.taskUUID) ?? []);
  }
  return results;
};
var executeWebSocket = async (tasks, transport, config, options) => {
  const isAsync = tasks.some((task) => task.deliveryMethod === "async");
  return isAsync ? executeWebSocketAsync(tasks, transport, config, options) : executeWebSocketSync(tasks, transport, config, options);
};
var toResultsArray = (response) => {
  if (Array.isArray(response?.data)) {
    return response.data;
  }
  if (Array.isArray(response)) {
    return response;
  }
  return [response];
};
var createItemDispatcher = (config, options) => {
  const seenCompleted = new Set;
  const lastProgressByKey = new Map;
  const keyFor = (item, index) => `${item.taskUUID}:${index}`;
  const safeCall = (cb, item) => {
    if (!cb) {
      return;
    }
    try {
      cb(item);
    } catch (error) {
      config.log.warn("User callback threw", error);
    }
  };
  const dispatch = (items) => {
    if (!options?.onResult && !options?.onProgress) {
      return;
    }
    items.forEach((item, idx) => {
      const key = keyFor(item, idx);
      const status = item.status;
      if (status === "success" || status === "error") {
        if (!seenCompleted.has(key)) {
          seenCompleted.add(key);
          safeCall(options?.onResult, item);
        }
        return;
      }
      if (typeof item.progress === "number") {
        if (lastProgressByKey.get(key) !== item.progress) {
          lastProgressByKey.set(key, item.progress);
          safeCall(options?.onProgress, item);
        }
      }
    });
  };
  return { dispatch };
};
var interruptibleSleep2 = async (ms, signal) => new Promise((resolve) => {
  if (!signal) {
    setTimeout(resolve, ms);
    return;
  }
  const timer = setTimeout(() => {
    signal.removeEventListener("abort", onAbort);
    resolve();
  }, ms);
  const onAbort = () => {
    clearTimeout(timer);
    resolve();
  };
  signal.addEventListener("abort", onAbort, { once: true });
});
var executeRest = async (tasks, transport, config, options) => {
  const signal = options?.signal;
  const response = await transport.sendRequest(tasks, { signal });
  const isAsync = tasks.some((task) => task.deliveryMethod === "async");
  if (!isAsync) {
    return toResultsArray(response);
  }
  if (Array.isArray(response?.errors) && response.errors.length > 0) {
    throw parseApiError({ errors: response.errors }, {
      taskType: tasks[0]?.taskType,
      model: tasks[0] ? tasks[0].model : undefined
    });
  }
  const dispatcher = createItemDispatcher(config, options);
  const expectedPerTask = new Map;
  const resultsPerTask = new Map;
  const restTasksByUUID = new Map;
  for (const task of tasks) {
    expectedPerTask.set(task.taskUUID, Number(task.numberResults) || 1);
    resultsPerTask.set(task.taskUUID, []);
    restTasksByUUID.set(task.taskUUID, task);
  }
  const pollTimeout = options?.timeout ?? config.pollTimeout;
  const startTime = Date.now();
  let delay = 1000;
  let pollNumber = 0;
  const pending = new Set(tasks.map((task) => task.taskUUID));
  config.log.request(`Polling ${pending.size} async task(s)`);
  while (pending.size > 0) {
    if (signal?.aborted) {
      throw createRunwareError("aborted", "Request aborted during polling");
    }
    if (Date.now() - startTime >= pollTimeout) {
      throw createRunwareError("timeout", `Polling timed out after ${pollTimeout}ms with ${pending.size} task(s) still processing`);
    }
    await interruptibleSleep2(delay, signal);
    if (signal?.aborted) {
      throw createRunwareError("aborted", "Request aborted during polling");
    }
    const uuids = Array.from(pending);
    const elapsed = Date.now() - startTime;
    const perPollTimeout = Math.max(1, Math.min(pollTimeout - elapsed, config.timeout));
    const polled = await Promise.all(uuids.map(async (taskUUID) => {
      const pollTasks = normalizeTasks("getResponse", { taskUUID });
      const pollOptions = { signal, timeout: perPollTimeout };
      const rawResponse = await transport.sendRequest(pollTasks, pollOptions);
      const pollResponse = rawResponse;
      return { taskUUID, response: pollResponse };
    }));
    for (const { taskUUID, response: pollResponse } of polled) {
      const items = Array.isArray(pollResponse?.data) ? pollResponse.data : [];
      const topLevelErrors = Array.isArray(pollResponse?.errors) ? pollResponse.errors : [];
      const itemErrors = items.filter((item) => item.status === "error").map((item) => ({ ...item.error ?? {}, taskUUID }));
      const errors = [...topLevelErrors, ...itemErrors];
      dispatcher.dispatch(items);
      if (errors.length > 0) {
        const origTask = restTasksByUUID.get(taskUUID);
        const parsed = parseApiError({ errors }, {
          taskType: origTask?.taskType,
          model: origTask ? origTask.model : undefined
        });
        if (!parsed.taskUUID) {
          parsed.taskUUID = taskUUID;
        }
        throw parsed;
      }
      const successItems = items.filter((item) => item.status === "success");
      resultsPerTask.set(taskUUID, successItems);
      const expected = expectedPerTask.get(taskUUID) ?? 1;
      if (successItems.length >= expected) {
        pending.delete(taskUUID);
      }
    }
    pollNumber += 1;
    delay = pollNumber === 1 ? 500 : Math.min(delay * 1.5, 1e4);
  }
  const results = [];
  for (const task of tasks) {
    results.push(...resultsPerTask.get(task.taskUUID) ?? []);
  }
  return results;
};
var execute = async (taskType, params, transport, config, options) => {
  const tasks = normalizeTasks(taskType, params);
  if (options?.signal?.aborted) {
    throw createRunwareError("aborted", "Request aborted before start");
  }
  if (options?.validate ?? config.validate) {
    await validateTasks(tasks, config);
  }
  if (transport.type === "websocket") {
    return executeWebSocket(tasks, transport, config, options);
  }
  return executeRest(tasks, transport, config, options);
};
var createClient = async (userConfig) => {
  const deps = isNodeJS() ? await createNodeDependencies() : createBrowserDependencies();
  const fullConfig = createConfig({
    ...userConfig,
    dependencies: { ...deps, ...userConfig.dependencies }
  });
  const transport = createTransport(fullConfig.transport, fullConfig);
  const registry = createRegistry({
    url: `${SCHEMAS_BASE_URL}/registry.json`,
    fetchImpl: fullConfig.dependencies?.fetch ?? globalThis.fetch,
    log: fullConfig.log,
    fallback: {
      models,
      architectureTaskTypes: { ...architectureTaskTypes, ...operationTaskTypes },
      modalityTaskTypes
    }
  });
  const connect = async () => {
    if (transport.type === "websocket") {
      await transport.connect();
    }
  };
  const disconnect = async () => {
    if (transport.type === "websocket") {
      await transport.disconnect();
    }
  };
  const isConnected = () => {
    return transport.type === "websocket" ? transport.isConnected() : true;
  };
  const normalizeModelParam = async (params) => {
    if (typeof params.model !== "string") {
      return params;
    }
    if (params.model.includes(":") && params.model.includes("@")) {
      return params;
    }
    const air = await registry.resolveModelAir(params.model);
    if (!air || air === params.model) {
      return params;
    }
    return { ...params, model: air };
  };
  const run = async (params, options) => {
    const encoded = await encodeLocalFiles(params);
    const normalized = await normalizeModelParam(encoded);
    const taskType = await resolveTaskType(undefined, normalized, registry);
    return execute(taskType, { deliveryMethod: "async", ...normalized }, transport, fullConfig, options);
  };
  const stream = async (params, options) => {
    if (typeof params.numberResults === "number" && params.numberResults > 1) {
      throw createRunwareError("notImplemented", "stream() with numberResults > 1 is not supported yet (backend does not emit resultIndex on stream chunks).");
    }
    const normalized = await normalizeModelParam(params);
    const taskType = await resolveTaskType(undefined, normalized, registry);
    const tasks = normalizeTasks(taskType, {
      ...normalized,
      deliveryMethod: "stream"
    });
    const shouldValidate = options?.validate ?? fullConfig.validate;
    const onStart = shouldValidate ? async () => validateTasks(tasks, fullConfig) : undefined;
    const timeout = options?.timeout;
    let signal = options?.signal;
    if (timeout != null) {
      const controller = new AbortController;
      const timer = setTimeout(() => controller.abort(createRunwareError("timeout", `Stream timed out after ${timeout}ms`)), timeout);
      if (signal) {
        if (signal.aborted) {
          clearTimeout(timer);
          controller.abort(signal.reason);
        } else {
          signal.addEventListener("abort", () => {
            clearTimeout(timer);
            controller.abort(signal.reason);
          }, { once: true });
        }
      }
      signal = controller.signal;
    }
    return createTextStream(tasks, fullConfig, signal, onStart);
  };
  const getResponse = async (params, options) => execute("getResponse", params, transport, fullConfig, options);
  const getTaskDetails = async (params, options) => execute("getTaskDetails", params, transport, fullConfig, options);
  const modelSearch = async (params, options) => execute("modelSearch", params, transport, fullConfig, options);
  const modelUpload = async (params, options) => execute("modelUpload", params, transport, fullConfig, options);
  const imageUpload = async (params, options) => {
    const encoded = await encodeLocalFiles(params);
    return execute("imageUpload", encoded, transport, fullConfig, options);
  };
  const mediaStorage = async (params, options) => {
    const encoded = await encodeLocalFiles(params);
    return execute("mediaStorage", encoded, transport, fullConfig, options);
  };
  const accountManagement = async (params, options) => execute("accountManagement", params, transport, fullConfig, options);
  const refreshRegistry = async () => registry.refresh();
  const content = createContentClient(fullConfig);
  return {
    connect,
    disconnect,
    isConnected,
    refreshRegistry,
    run,
    stream,
    getResponse,
    getTaskDetails,
    modelSearch,
    modelUpload,
    imageUpload,
    mediaStorage,
    accountManagement,
    content
  };
};
export {
  parseSseLine,
  parseApiError,
  operationTaskTypes,
  models,
  modalityTaskTypes,
  isRunwareError,
  fileToDataURI,
  createRunwareError,
  createRegistry,
  createLogger,
  createClient,
  clearValidatorCache,
  architectureTaskTypes,
  RunwareError
};

//# debugId=EA953F0D359D479464756E2164756E21
